// aws-cdk/test/api/fake-cloudformation-stack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FakeCloudformationStack = void 0;
/* eslint-disable import/order */
const aws_sdk_1 = require("aws-sdk");
const cloudformation_1 = require("../../lib/api/util/cloudformation");
const util_1 = require("../util");
const stack_status_1 = require("../../lib/api/util/cloudformation/stack-status");
class FakeCloudformationStack extends cloudformation_1.CloudFormationStack {
    constructor(props) {
        const cfnMock = (0, util_1.instanceMockFrom)(aws_sdk_1.CloudFormation);
        super(cfnMock, props.stackName);
        this.cfnMock = cfnMock;
        this.props = props;
        this.__template = {};
    }
    setTemplate(template) {
        this.__template = template;
    }
    async template() {
        return Promise.resolve(this.__template);
    }
    get exists() {
        return this.props.stackId !== undefined;
    }
    get stackStatus() {
        const status = this.props.stackStatus ?? 'UPDATE_COMPLETE';
        return new stack_status_1.StackStatus(status, 'The test said so');
    }
    get stackId() {
        if (!this.props.stackId) {
            throw new Error('Cannot retrieve stackId from a non-existent stack');
        }
        return this.props.stackId;
    }
    get outputs() {
        return {};
    }
}
exports.FakeCloudformationStack = FakeCloudformationStack;
