// @aws-amplify/backend/lib/engine/custom_outputs_accumulator.js
import { customOutputKey, } from '@aws-amplify/backend-output-schemas';
import { Lazy } from 'aws-cdk-lib';
import { AmplifyUserError, ObjectAccumulatorPropertyAlreadyExistsError, ObjectAccumulatorVersionMismatchError, } from '@aws-amplify/platform-core';
/**
 * Accumulates custom outputs as they're added to the backend.
 */
export class CustomOutputsAccumulator {
    outputStorageStrategy;
    clientConfigAccumulator;
    hasBackendOutputEntry = false;
    /**
     * Creates custom outputs accumulator.
     */
    constructor(outputStorageStrategy, clientConfigAccumulator) {
        this.outputStorageStrategy = outputStorageStrategy;
        this.clientConfigAccumulator = clientConfigAccumulator;
    }
    addOutput = (clientConfigPart) => {
        try {
            this.clientConfigAccumulator.accumulate(clientConfigPart);
        }
        catch (error) {
            if (error instanceof ObjectAccumulatorPropertyAlreadyExistsError) {
                throw new AmplifyUserError('OutputEntryAlreadyExistsError', {
                    message: `Output entry with key ${error.key} already exists`,
                    resolution: "Check if 'backend.addOutput' is called multiple times with overlapping inputs",
                }, error);
            }
            if (error instanceof ObjectAccumulatorVersionMismatchError) {
                throw new AmplifyUserError('VersionMismatchError', {
                    message: `Conflicting versions of client configuration found.`,
                    resolution: "Ensure that the version specified in 'backend.addOutput' is consistent" +
                        ' and is same as the one used for generating the client config',
                }, error);
            }
            throw error;
        }
        this.ensureBackendOutputEntry();
    };
    ensureBackendOutputEntry = () => {
        if (this.hasBackendOutputEntry) {
            return;
        }
        this.outputStorageStrategy.addBackendOutputEntry(customOutputKey, {
            version: '1',
            payload: {
                customOutputs: Lazy.string({
                    produce: () => {
                        return JSON.stringify(this.clientConfigAccumulator.getAccumulatedObject());
                    },
                }),
            },
        });
        this.hasBackendOutputEntry = true;
    };
}
