// @aws-amplify/codegen-ui/dist/lib/generate-form-definition/generate-form-definition.js
"use strict";
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateFormDefinition = void 0;
const helpers_1 = require("./helpers");
const errors_1 = require("../errors");
/**
 * Helper that turns the StudioForm model into definition that can be used to render
 * UI Components in the customer project and shapes representing UI Components in Studio.
 * @param form StudioForm, converted from the API shape.
 * @param modelInfo (Optional) holds type information about the DataStore model fields being represented.
 * @returns a definition that translates to rendered JSX elements.
 */
function generateFormDefinition({ form, dataSchema, featureFlags, }) {
    const formDefinition = {
        form: { layoutStyle: (0, helpers_1.mapStyles)(form.style), labelDecorator: form.labelDecorator || 'none' },
        elements: {},
        buttons: {
            buttonConfigs: {},
            position: '',
            buttonMatrix: [],
        },
        elementMatrix: [],
    };
    let modelFieldsConfigs = {};
    if (form.dataType.dataSourceType !== 'Custom') {
        if (!dataSchema) {
            throw new errors_1.InvalidInputError(`Data schema is missing for form of data source type ${form.dataType.dataSourceType}`);
        }
        modelFieldsConfigs = (0, helpers_1.mapModelFieldsConfigs)({
            dataSchema,
            formDefinition,
            dataTypeName: form.dataType.dataTypeName,
            formActionType: form.formActionType,
            featureFlags,
        });
    }
    const elementQueue = Object.entries(form.fields)
        .map(([fieldName, fieldConfig]) => ({
        name: fieldName,
        position: typeof fieldConfig === 'object' && 'position' in fieldConfig ? fieldConfig.position : undefined,
        excluded: typeof fieldConfig === 'object' && 'excluded' in fieldConfig ? fieldConfig.excluded : false,
    }))
        .concat(Object.entries(form.sectionalElements).map(([elementName, elementConfig]) => ({
        name: elementName,
        position: 'position' in elementConfig ? elementConfig.position : undefined,
        excluded: 'excluded' in elementConfig ? elementConfig.excluded : false,
    })));
    (0, helpers_1.mapElementMatrix)({ elementQueue, formDefinition });
    (0, helpers_1.mapElements)({ form, formDefinition, modelFieldsConfigs });
    formDefinition.buttons = (0, helpers_1.mapButtons)(form.formActionType, form.cta);
    return formDefinition;
}
exports.generateFormDefinition = generateFormDefinition;
