// @aws-amplify/graphql-docs-generator/lib/cli.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = void 0;
const yargs = require("yargs");
const logger_1 = require("./logger");
const index_1 = require("./index");
// Make sure unhandled errors in async code are propagated correctly
process.on('unhandledRejection', error => {
    throw error;
});
process.on('uncaughtException', handleError);
function handleError(error) {
    (0, logger_1.logError)(error);
    process.exit(1);
}
function run(argv) {
    // tslint:disable
    yargs
        .command('$0', 'Generate graphql operations for the provided introspection schema', {
        schema: {
            demand: true,
            describe: 'GraphQL introspection or SDL schema',
            normalize: true,
        },
        maxDepth: {
            demand: true,
            default: 2,
            normalize: true,
            type: 'number',
        },
        typenameIntrospection: {
            default: true,
            type: 'boolean',
        },
    }, (argv) => __awaiter(this, void 0, void 0, function* () {
        (0, index_1.generateGraphQLDocuments)(argv.schema, { maxDepth: argv.maxDepth, typenameIntrospection: argv.typenameIntrospection });
    }))
        .help()
        .version()
        .strict().argv;
}
exports.run = run;
