// @aws-amplify/sandbox/lib/files_changes_tracker.js
import { glob } from 'glob';
/**
 * This class accumulates information about files that have changed.
 */
export class FilesChangesTracker {
    snapshot;
    /**
     * Creates a new file changes tracker.
     */
    constructor(snapshot) {
        this.snapshot = snapshot;
    }
    trackFileChange = (path) => {
        this.snapshot.didAnyFileChangeSinceStart = true;
        if (path.endsWith('.ts') || path.endsWith('.tsx')) {
            this.snapshot.didAnyTypeScriptFileChangeSinceLastSnapshot = true;
        }
    };
    getAndResetSnapshot = () => {
        // shallow copy is ok for 1 dim data structure
        const currentSnapshot = { ...this.snapshot };
        this.resetSnapshot();
        return currentSnapshot;
    };
    resetSnapshot = () => {
        this.snapshot.didAnyTypeScriptFileChangeSinceLastSnapshot = false;
    };
}
/**
 * Creates a new instance of files changes tracker.
 */
export const createFilesChangesTracker = async (rootPath) => {
    const tsFiles = await glob('**/*.{ts,tsx}', {
        ignore: 'node_modules/**',
        cwd: rootPath,
    });
    return new FilesChangesTracker({
        didAnyFileChangeSinceStart: false,
        hadTypeScriptFilesAtStart: tsFiles.length > 0,
        didAnyTypeScriptFileChangeSinceLastSnapshot: false,
    });
};
