// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-secret/sandbox_secret_get_command.js
import { format, printer } from '@aws-amplify/cli-core';
/**
 * Command to get sandbox secret.
 */
export class SandboxSecretGetCommand {
    sandboxIdResolver;
    secretClient;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Get sandbox secret command.
     */
    constructor(sandboxIdResolver, secretClient) {
        this.sandboxIdResolver = sandboxIdResolver;
        this.secretClient = secretClient;
        this.command = 'get <secret-name>';
        this.describe = 'Get a sandbox secret';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const sandboxBackendIdentifier = await this.sandboxIdResolver.resolve(args.identifier);
        const secret = await this.secretClient.getSecret(sandboxBackendIdentifier, {
            name: args.secretName,
        });
        printer.print(format.record(secret));
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs
            .positional('secret-name', {
            describe: 'Name of the secret to get',
            type: 'string',
            demandOption: true,
        })
            .help();
    };
}
