// @aws-amplify/codegen-ui/dist/lib/types/bindings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StudioComponentPropertyBindingType = exports.StudioComponentPropertyType = void 0;
/**
 * These are the primitive value types
 */
var StudioComponentPropertyType;
(function (StudioComponentPropertyType) {
    StudioComponentPropertyType["String"] = "String";
    StudioComponentPropertyType["Number"] = "Number";
    StudioComponentPropertyType["Boolean"] = "Boolean";
    StudioComponentPropertyType["Date"] = "Date";
})(StudioComponentPropertyType = exports.StudioComponentPropertyType || (exports.StudioComponentPropertyType = {}));
/**
 * These are the types of data binding
 */
var StudioComponentPropertyBindingType;
(function (StudioComponentPropertyBindingType) {
    StudioComponentPropertyBindingType["Data"] = "Data";
    StudioComponentPropertyBindingType["Storage"] = "Storage";
    StudioComponentPropertyBindingType["Event"] = "Event";
})(StudioComponentPropertyBindingType = exports.StudioComponentPropertyBindingType || (exports.StudioComponentPropertyBindingType = {}));
