// @aws-amplify/codegen-ui/dist/lib/types/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StudioGenericEvent = void 0;
var StudioGenericEvent;
(function (StudioGenericEvent) {
    StudioGenericEvent["onClick"] = "onClick";
    StudioGenericEvent["onDoubleClick"] = "onDoubleClick";
    StudioGenericEvent["onMouseDown"] = "onMouseDown";
    StudioGenericEvent["onMouseEnter"] = "onMouseEnter";
    StudioGenericEvent["onMouseLeave"] = "onMouseLeave";
    StudioGenericEvent["onMouseMove"] = "onMouseMove";
    StudioGenericEvent["onMouseOut"] = "onMouseOut";
    StudioGenericEvent["onMouseOver"] = "onMouseOver";
    StudioGenericEvent["onMouseUp"] = "onMouseUp";
    StudioGenericEvent["onChange"] = "onChange";
    StudioGenericEvent["onInput"] = "onInput";
    StudioGenericEvent["onFocus"] = "onFocus";
    StudioGenericEvent["onBlur"] = "onBlur";
    StudioGenericEvent["onKeyDown"] = "onKeyDown";
    StudioGenericEvent["onKeyPress"] = "onKeyPress";
    StudioGenericEvent["onKeyUp"] = "onKeyUp";
    StudioGenericEvent["onSubmit"] = "onSubmit";
})(StudioGenericEvent = exports.StudioGenericEvent || (exports.StudioGenericEvent = {}));
