// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-delete/sandbox_delete_command.js
import { AmplifyPrompter } from '@aws-amplify/cli-core';
/**
 * Command that deletes the sandbox environment.
 */
export class SandboxDeleteCommand {
    sandboxFactory;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Deletes sandbox environment.
     */
    constructor(sandboxFactory) {
        this.sandboxFactory = sandboxFactory;
        this.command = 'delete';
        this.describe = 'Deletes sandbox environment';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        let isConfirmed = args.yes;
        if (!isConfirmed) {
            isConfirmed = await AmplifyPrompter.yesOrNo({
                message: "Are you sure you want to delete all the resources in your sandbox environment (This can't be undone)?",
            });
        }
        if (isConfirmed) {
            await (await this.sandboxFactory.getInstance()).delete({ identifier: args.identifier, profile: args.profile });
        }
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs.option('yes', {
            describe: 'Do not ask for confirmation before deleting the sandbox environment',
            type: 'boolean',
            array: false,
            alias: 'y',
        });
    };
}
