// @aws-amplify/backend-data/lib/app_sync_policy_generator.js
import { Stack } from 'aws-cdk-lib';
import { Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
/**
 * Generates policies for accessing an AppSync GraphQL API
 */
export class AppSyncPolicyGenerator {
    graphqlApi;
    stack;
    policyPrefix = 'GraphqlAccessPolicy';
    policyCount = 1;
    /**
     * Initialize with the GraphqlAPI that the policies will be scoped to
     */
    constructor(graphqlApi) {
        this.graphqlApi = graphqlApi;
        this.stack = Stack.of(graphqlApi);
    }
    /**
     * Generates a policy that grants GraphQL data-plane access to the provided actions
     *
     * The naming is a bit wonky here because the IAM action is always "appsync:GraphQL".
     * The input "action" maps to the "type" in the resource name part of the ARN which is "Query", "Mutation" or "Subscription"
     */
    generateGraphqlAccessPolicy(actions) {
        const resources = actions
            // convert from actions to GraphQL Type
            .map((action) => actionToTypeMap[action])
            // convert Type to resourceName
            .map((type) => [this.graphqlApi.arn, 'types', type, '*'].join('/'));
        return new Policy(this.stack, `${this.policyPrefix}${this.policyCount++}`, {
            statements: [
                new PolicyStatement({
                    actions: ['appsync:GraphQL'],
                    resources,
                }),
            ],
        });
    }
}
const actionToTypeMap = {
    query: 'Query',
    mutate: 'Mutation',
    listen: 'Subscription',
};
