// @aws-amplify/graphql-generation-transformer/lib/utils/tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createResponseTypeTool = void 0;
const graphql_json_schema_type_1 = require("./graphql-json-schema-type");
const createResponseTypeTool = (field, ctx) => {
    const { type } = field;
    const schema = (0, graphql_json_schema_type_1.generateJSONSchemaFromTypeNode)(type, ctx);
    const boxedSchema = {
        type: 'object',
        properties: {
            value: schema,
        },
        required: ['value'],
    };
    const description = 'Generate a response type for the given field.';
    const tools = [
        {
            toolSpec: {
                name: 'responseType',
                description,
                inputSchema: {
                    json: boxedSchema,
                },
            },
        },
    ];
    const toolChoice = { tool: { name: 'responseType' } };
    const toolConfig = { tools, toolChoice };
    return toolConfig;
};
exports.createResponseTypeTool = createResponseTypeTool;
