// @aws-amplify/backend-output-schemas/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unifiedBackendOutputSchema = exports.aiConversationOutputKey = exports.functionOutputKey = exports.storageOutputKey = exports.graphqlOutputKey = exports.authOutputKey = exports.customOutputKey = exports.platformOutputKey = void 0;
const zod_1 = require("zod");
const index_js_1 = require("./auth/index.js");
const index_js_2 = require("./graphql/index.js");
const index_js_3 = require("./storage/index.js");
const index_js_4 = require("./stack/index.js");
const custom_1 = require("./custom");
const index_js_5 = require("./function/index.js");
const index_js_6 = require("./ai/conversation/index.js");
/**
 * The auth, graphql and storage exports here are duplicated from the submodule exports in the package.json file
 * This is because these types need to be consumed in CDK Constructs that may be JSII classes. JSII only supports
 * CommonJS modules which can't interpret submodule exports
 */
/**
 * ---------- Platform exports ----------
 */
exports.platformOutputKey = 'AWS::Amplify::Platform';
__exportStar(require("./platform/index.js"), exports);
/**
 * ---------- Custom exports ----------
 */
exports.customOutputKey = 'AWS::Amplify::Custom';
__exportStar(require("./custom/index.js"), exports);
/**
 * ---------- Auth exports ----------
 */
/**
 * re-export the auth output schema
 */
__exportStar(require("./auth/index.js"), exports);
/**
 * Expected key that auth output is stored under
 */
exports.authOutputKey = 'AWS::Amplify::Auth';
/**
 * ---------- GraphQL exports ----------
 */
/**
 * re-export the graphql output schema
 */
__exportStar(require("./graphql/index.js"), exports);
/**
 * Expected key that graphql output is stored under
 */
exports.graphqlOutputKey = 'AWS::Amplify::GraphQL';
/**
 * ---------- Storage exports ----------
 */
/**
 * re-export the storage output schema
 */
__exportStar(require("./storage/index.js"), exports);
/**
 * Expected key that storage output is stored under
 */
exports.storageOutputKey = 'AWS::Amplify::Storage';
/**
 * ---------- Function exports ----------
 */
/**
 * re-export the function output schema
 */
__exportStar(require("./function/index.js"), exports);
/**
 * Expected key that function output is stored under
 */
exports.functionOutputKey = 'AWS::Amplify::Function';
/**
 * ---------- AI conversation exports ----------
 */
/**
 * re-export the AI conversation output schema
 */
__exportStar(require("./ai/conversation/index.js"), exports);
/**
 * Expected key that AI conversation output is stored under
 */
exports.aiConversationOutputKey = 'AWS::Amplify::AI::Conversation';
/**
 * ---------- Unified exports ----------
 */
/**
 * Defines the unified expected shape of Amplify backend output.
 * As new constructs are added that need to contribute backend output, entries should be added here so that client config generation is aware of these outputs
 */
exports.unifiedBackendOutputSchema = zod_1.z.object({
    [exports.platformOutputKey]: index_js_4.versionedStackOutputSchema.optional(),
    [exports.authOutputKey]: index_js_1.versionedAuthOutputSchema.optional(),
    [exports.graphqlOutputKey]: index_js_2.versionedGraphqlOutputSchema.optional(),
    [exports.storageOutputKey]: index_js_3.versionedStorageOutputSchema.optional(),
    [exports.customOutputKey]: custom_1.versionedCustomOutputSchema.optional(),
    [exports.functionOutputKey]: index_js_5.versionedFunctionOutputSchema.optional(),
    [exports.aiConversationOutputKey]: index_js_6.versionedAIConversationOutputSchema.optional(),
});
