// @aws-amplify/graphql-transformer-core/lib/utils/defaultSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SCHEMA_DEFINITION = exports.DEFAULT_SUBSCRIPTION_OPERATION = exports.DEFAULT_MUTATION_OPERATION = exports.DEFAULT_QUERY_OPERATION = void 0;
const graphql_1 = require("graphql");
exports.DEFAULT_QUERY_OPERATION = {
    kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
    operation: 'query',
    type: {
        kind: graphql_1.Kind.NAMED_TYPE,
        name: {
            kind: graphql_1.Kind.NAME,
            value: 'Query',
        },
    },
};
exports.DEFAULT_MUTATION_OPERATION = {
    kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
    operation: 'mutation',
    type: {
        kind: graphql_1.Kind.NAMED_TYPE,
        name: {
            kind: graphql_1.Kind.NAME,
            value: 'Mutation',
        },
    },
};
exports.DEFAULT_SUBSCRIPTION_OPERATION = {
    kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
    operation: 'subscription',
    type: {
        kind: graphql_1.Kind.NAMED_TYPE,
        name: {
            kind: graphql_1.Kind.NAME,
            value: 'Subscription',
        },
    },
};
exports.DEFAULT_SCHEMA_DEFINITION = {
    kind: graphql_1.Kind.SCHEMA_DEFINITION,
    directives: [],
    operationTypes: [exports.DEFAULT_QUERY_OPERATION, exports.DEFAULT_MUTATION_OPERATION, exports.DEFAULT_SUBSCRIPTION_OPERATION],
};
