// @aws-amplify/cli-core/lib/format/format.js
import * as os from 'node:os';
import { blue, bold, cyan, dim, green, grey, magenta, red, underline, yellow, } from 'kleur/colors';
import { AmplifyFault } from '@aws-amplify/platform-core';
import { getPackageManagerRunnerName } from '../package-manager-controller/get_package_manager_name.js';
/**
 * Formats various inputs into single string.
 */
export class Format {
    packageManagerRunnerName;
    /**
     * Initialize with a package manager runner name. Defaults to attempting to load a value derived from environment variables.
     */
    constructor(packageManagerRunnerName = getPackageManagerRunnerName()) {
        this.packageManagerRunnerName = packageManagerRunnerName;
    }
    normalizeAmpxCommand = (command) => {
        if (command.length === 0) {
            throw new AmplifyFault('InvalidFormatFault', {
                message: 'The command must be non-empty',
            });
        }
        return cyan(`${this.packageManagerRunnerName} ampx ${command}`);
    };
    error = (error) => {
        if (error instanceof Error) {
            const message = red(`${error.name}: ${error.message}`);
            if (error.cause) {
                return message + os.EOL + this.error(error.cause);
            }
            return message;
        }
        else if (typeof error === 'string') {
            return red(error);
        }
        try {
            return red(JSON.stringify(error, null, 2));
        }
        catch (e) {
            return red('Unknown error') + os.EOL + this.error(e);
        }
    };
    note = (message) => grey(message);
    command = (command) => cyan(command);
    highlight = (command) => cyan(command);
    success = (message) => green(message);
    sectionHeader = (header) => bold(blue(header));
    bold = (message) => bold(message);
    dim = (message) => dim(message);
    link = (link) => underline(blue(link));
    list = (lines) => lines.map((line) => ` - ${line}`).join(os.EOL);
    indent = (message) => {
        if (message === '') {
            throw new Error('Message cannot be empty');
        }
        const spaces = '  '; // Two spaces for indentation
        return message
            .split(os.EOL)
            .map((line) => `${spaces}${line}`)
            .join(os.EOL);
    };
    record = (record) => Object.entries(record)
        .map(([key, value]) => `${key}: ${String(value)}`)
        .join(os.EOL);
    color = (message, colorName) => colors[colorName](message);
}
// Map to connect colorName to kleur color
const colors = {
    Green: green,
    Yellow: yellow,
    Blue: blue,
    Magenta: magenta,
    Cyan: cyan,
};
export const colorNames = [
    'Green',
    'Yellow',
    'Blue',
    'Magenta',
    'Cyan',
];
export const format = new Format();
