// @aws-amplify/codegen-ui-react/dist/lib/amplify-ui-renderers/collection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const codegen_ui_1 = require("@aws-amplify/codegen-ui");
const typescript_1 = __importStar(require("typescript"));
const react_component_renderer_1 = require("../react-component-renderer");
const react_component_render_helper_1 = require("../react-component-render-helper");
const imports_1 = require("../imports");
const amplify_js_versioning_1 = require("../helpers/amplify-js-versioning");
class CollectionRenderer extends react_component_renderer_1.ReactComponentRenderer {
    constructor(component, componentMetadata, importCollection, renderConfig, parent) {
        super(component, componentMetadata, importCollection, parent);
        this.componentMetadata = componentMetadata;
        this.importCollection = importCollection;
        this.renderConfig = renderConfig;
        this.parent = parent;
    }
    renderElement(renderChildren) {
        var _a, _b, _c, _d;
        this.addKeyPropertyToChildren((_a = this.component.children) !== null && _a !== void 0 ? _a : []);
        const childrenJsx = this.component.children ? renderChildren((_b = this.component.children) !== null && _b !== void 0 ? _b : []) : [];
        const element = this.getCollectionElement(childrenJsx, (_c = this.renderConfig.apiConfiguration) === null || _c === void 0 ? void 0 : _c.dataApi);
        this.importCollection.addImport('@aws-amplify/ui-react', this.component.componentType);
        if (((_d = this.renderConfig.apiConfiguration) === null || _d === void 0 ? void 0 : _d.dataApi) === 'GraphQL') {
            const mappedImport = (0, amplify_js_versioning_1.getAmplifyJSAPIImport)(this.renderConfig.dependencies);
            this.importCollection.addMappedImport(mappedImport);
            this.importCollection.addMappedImport(imports_1.ImportValue.PAGINATION, imports_1.ImportValue.PLACEHOLDER);
        }
        return element;
    }
    renderCollectionOpeningElement(itemsVariableName) {
        var _a;
        const propsArray = Object.entries(this.component.properties).reduce((acc, [key, value]) => {
            var _a;
            if (((_a = this.renderConfig.apiConfiguration) === null || _a === void 0 ? void 0 : _a.dataApi) === 'GraphQL') {
                if (key === 'isPaginated' || key === 'itemsPerPage') {
                    return acc;
                }
            }
            return [...acc, (0, react_component_render_helper_1.buildOpeningElementProperties)(this.componentMetadata, value, key)];
        }, []);
        let itemsAttribute;
        if (((_a = this.renderConfig.apiConfiguration) === null || _a === void 0 ? void 0 : _a.dataApi) === 'GraphQL') {
            propsArray.push(typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('itemsPerPage'), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createIdentifier('pageSize'))), typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('isPaginated'), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createBinaryExpression(typescript_1.factory.createPrefixUnaryExpression(typescript_1.default.SyntaxKind.ExclamationToken, typescript_1.factory.createIdentifier('isApiPagination')), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.AmpersandAmpersandToken), typescript_1.factory.createIdentifier('isPaginated')))));
            itemsAttribute = typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('items'), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createBinaryExpression(typescript_1.factory.createIdentifier('itemsProp'), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.BarBarToken), typescript_1.factory.createParenthesizedExpression(typescript_1.factory.createConditionalExpression(typescript_1.factory.createIdentifier('loading'), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.QuestionToken), typescript_1.factory.createCallExpression(typescript_1.factory.createPropertyAccessExpression(typescript_1.factory.createNewExpression(typescript_1.factory.createIdentifier('Array'), undefined, [
                typescript_1.factory.createIdentifier('pageSize'),
            ]), typescript_1.factory.createIdentifier('fill')), undefined, [typescript_1.factory.createObjectLiteralExpression([], false)]), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.ColonToken), typescript_1.factory.createIdentifier('items'))))));
        }
        else {
            itemsAttribute = typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('items'), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createBinaryExpression(typescript_1.factory.createIdentifier(itemsVariableName || 'items'), typescript_1.factory.createToken(typescript_1.SyntaxKind.BarBarToken), typescript_1.factory.createArrayLiteralExpression([], false))));
        }
        propsArray.push(itemsAttribute);
        this.addPropsSpreadAttributes(propsArray);
        return typescript_1.factory.createJsxOpeningElement(typescript_1.factory.createIdentifier(this.component.componentType), undefined, typescript_1.factory.createJsxAttributes(propsArray));
    }
    addKeyPropertyToChildren(children) {
        children.forEach((child) => {
            var _a, _b, _c;
            if ('key' in child.properties) {
                return;
            }
            let keys = ['id'];
            if ((0, codegen_ui_1.isStudioComponentWithCollectionProperties)(this.component)) {
                const firstCollectionProp = Object.entries((_a = this.component.collectionProperties) !== null && _a !== void 0 ? _a : {})[0];
                if (firstCollectionProp) {
                    const [, { model }] = firstCollectionProp;
                    const primaryKeys = (_c = (_b = this.componentMetadata.dataSchemaMetadata) === null || _b === void 0 ? void 0 : _b.models[model]) === null || _c === void 0 ? void 0 : _c.primaryKeys;
                    if (primaryKeys) {
                        keys = primaryKeys;
                    }
                }
            }
            const bindingProperties = keys.map((key) => ({
                collectionBindingProperties: {
                    property: '',
                    field: key,
                },
            }));
            // eslint-disable-next-line no-param-reassign
            child.properties.key = bindingProperties.length === 1 ? bindingProperties[0] : { concat: bindingProperties };
        });
    }
    findItemsVariableName() {
        var _a;
        if ((0, codegen_ui_1.isStudioComponentWithCollectionProperties)(this.component)) {
            const collectionProps = Object.entries((_a = this.component.collectionProperties) !== null && _a !== void 0 ? _a : {});
            return collectionProps.length > 0 ? collectionProps[0][0] : undefined;
        }
        return undefined;
    }
    renderItemArrowFunctionExpr(childrenJsx) {
        var _a;
        if (((_a = this.renderConfig.apiConfiguration) === null || _a === void 0 ? void 0 : _a.dataApi) === 'GraphQL') {
            return typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createArrowFunction(undefined, undefined, [
                typescript_1.factory.createParameterDeclaration(undefined, undefined, undefined, typescript_1.factory.createIdentifier('item'), undefined, undefined, undefined),
                typescript_1.factory.createParameterDeclaration(undefined, undefined, undefined, typescript_1.factory.createIdentifier('index'), undefined, undefined, undefined),
            ], undefined, typescript_1.factory.createToken(typescript_1.default.SyntaxKind.EqualsGreaterThanToken), typescript_1.factory.createBlock([
                typescript_1.factory.createIfStatement(typescript_1.factory.createIdentifier('loading'), typescript_1.factory.createBlock([
                    typescript_1.factory.createReturnStatement(typescript_1.factory.createParenthesizedExpression(typescript_1.factory.createJsxSelfClosingElement(typescript_1.factory.createIdentifier('Placeholder'), undefined, typescript_1.factory.createJsxAttributes([
                        typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('key'), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createIdentifier('index'))),
                        typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('size'), typescript_1.factory.createStringLiteral('large')),
                    ])))),
                ], true), undefined),
                typescript_1.factory.createReturnStatement(typescript_1.factory.createParenthesizedExpression(childrenJsx[0])),
            ], true)));
        }
        return typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createArrowFunction(undefined, undefined, [
            typescript_1.factory.createParameterDeclaration(undefined, undefined, undefined, typescript_1.factory.createIdentifier('item'), undefined, undefined, undefined),
            typescript_1.factory.createParameterDeclaration(undefined, undefined, undefined, typescript_1.factory.createIdentifier('index'), undefined, undefined, undefined),
        ], undefined, typescript_1.factory.createToken(typescript_1.SyntaxKind.EqualsGreaterThanToken), typescript_1.factory.createParenthesizedExpression(childrenJsx[0])));
    }
    getCollectionElement(childrenJsx, dataApi = 'DataStore') {
        const arrowFuncExpr = this.renderItemArrowFunctionExpr(childrenJsx);
        const itemsVariableName = this.findItemsVariableName();
        if (dataApi === 'GraphQL') {
            const attributesObj = [
                { name: 'currentPage', initialValue: 'pageIndex' },
                { name: 'totalPages', initialValue: 'maxViewed' },
                { name: 'hasMorePages', initialValue: 'hasMorePages' },
                { name: 'onNext', initialValue: 'handleNextPage' },
                { name: 'onPrevious', initialValue: 'handlePreviousPage' },
                { name: 'onChange', initialValue: 'jumpToPage' },
            ];
            const attributes = [];
            attributesObj.forEach((attribute) => {
                attributes.push(typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier(attribute.name), typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createIdentifier(attribute.initialValue))));
            });
            return typescript_1.factory.createJsxElement(typescript_1.factory.createJsxOpeningElement(typescript_1.factory.createIdentifier('div'), undefined, typescript_1.factory.createJsxAttributes([])), [
                typescript_1.factory.createJsxElement(this.renderCollectionOpeningElement(itemsVariableName), [arrowFuncExpr], typescript_1.factory.createJsxClosingElement(typescript_1.factory.createIdentifier(this.component.componentType))),
                typescript_1.factory.createJsxExpression(undefined, typescript_1.factory.createBinaryExpression(typescript_1.factory.createBinaryExpression(typescript_1.factory.createIdentifier('isApiPagination'), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.AmpersandAmpersandToken), typescript_1.factory.createIdentifier('isPaginated')), typescript_1.factory.createToken(typescript_1.default.SyntaxKind.AmpersandAmpersandToken), typescript_1.factory.createJsxSelfClosingElement(typescript_1.factory.createIdentifier('Pagination'), undefined, typescript_1.factory.createJsxAttributes(attributes)))),
            ], typescript_1.factory.createJsxClosingElement(typescript_1.factory.createIdentifier('div')));
        }
        return typescript_1.factory.createJsxElement(this.renderCollectionOpeningElement(itemsVariableName), [arrowFuncExpr], typescript_1.factory.createJsxClosingElement(typescript_1.factory.createIdentifier(this.component.componentType)));
    }
}
exports.default = CollectionRenderer;
