// @aws-amplify/graphql-transformer-core/lib/appsync-function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppSyncFunctionConfiguration = void 0;
const aws_appsync_1 = require("aws-cdk-lib/aws-appsync");
const constructs_1 = require("constructs");
const template_asset_1 = require("./cdk-compat/template-asset");
const utils_1 = require("./utils");
class AppSyncFunctionConfiguration extends constructs_1.Construct {
    constructor(scope, id, props) {
        var _a;
        super(scope, id);
        const requestTemplate = props.requestMappingTemplate.bind(this, props.api.assetProvider);
        const responseTemplate = props.responseMappingTemplate.bind(this, props.api.assetProvider);
        this.function = new aws_appsync_1.CfnFunctionConfiguration(this, `${id}.AppSyncFunction`, {
            name: id,
            apiId: props.api.apiId,
            functionVersion: '2018-05-29',
            description: props.description,
            dataSourceName: props.dataSource instanceof aws_appsync_1.BaseDataSource ? props.dataSource.ds.attrName : props.dataSource,
            ...(props.requestMappingTemplate instanceof template_asset_1.InlineTemplate
                ? { requestMappingTemplate: requestTemplate }
                : { requestMappingTemplateS3Location: requestTemplate }),
            ...(props.responseMappingTemplate instanceof template_asset_1.InlineTemplate
                ? { responseMappingTemplate: responseTemplate }
                : { responseMappingTemplateS3Location: responseTemplate }),
        });
        (0, utils_1.setResourceName)(this.function, { name: id });
        props.api.addSchemaDependency(this.function);
        if (props.dataSource instanceof aws_appsync_1.BackedDataSource) {
            this.function.addDependency((_a = props.dataSource) === null || _a === void 0 ? void 0 : _a.ds);
        }
        this.arn = this.function.attrFunctionArn;
        this.functionId = this.function.attrFunctionId;
    }
}
exports.AppSyncFunctionConfiguration = AppSyncFunctionConfiguration;
