// @aws-amplify/graphql-relational-transformer/lib/belongs-to/belongs-to-directive-ddb-references-transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BelongsToDirectiveDDBReferencesTransformer = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const ddb_references_generator_1 = require("../resolver/ddb-references-generator");
const resolvers_1 = require("../resolvers");
const utils_1 = require("../utils");
class BelongsToDirectiveDDBReferencesTransformer {
    constructor() {
        this.dbType = graphql_transformer_core_1.DDB_DB_TYPE;
        this.prepare = (context, config) => {
            var _a, _b;
            const modelName = config.object.name.value;
            (0, resolvers_1.setFieldMappingResolverReference)(context, (_b = (_a = config.relatedType) === null || _a === void 0 ? void 0 : _a.name) === null || _b === void 0 ? void 0 : _b.value, modelName, config.field.name.value);
            (0, utils_1.registerHasOneForeignKeyMappings)({
                transformParameters: context.transformParameters,
                resourceHelper: context.resourceHelper,
                thisTypeName: modelName,
                thisFieldName: config.field.name.value,
                relatedType: config.relatedType,
            });
        };
        this.transformSchema = (context, config) => { };
        this.generateResolvers = (context, config) => {
            new ddb_references_generator_1.DDBRelationalReferencesResolverGenerator().makeBelongsToGetItemConnectionWithKeyResolver(config, context);
        };
        this.validate = (context, config) => {
            (0, utils_1.ensureReferencesArray)(config);
            (0, utils_1.validateChildReferencesFields)(config, context);
            (0, utils_1.validateReferencesRelationalFieldNullability)(config);
            config.referenceNodes = (0, utils_1.getBelongsToReferencesNodes)(config, context);
            (0, utils_1.validateReferencesBidirectionality)(config);
        };
    }
}
exports.BelongsToDirectiveDDBReferencesTransformer = BelongsToDirectiveDDBReferencesTransformer;
