// @aws-amplify/codegen-ui-react/dist/lib/utils-file-functions/hooks/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constantsString = exports.DATASTORE_QUERY_BY_ID_ERROR = exports.ACTION_DATASTORE_UPDATE_FINISHED = exports.ACTION_DATASTORE_UPDATE_STARTED = exports.EVENT_ACTION_DATASTORE_UPDATE = exports.ACTION_DATASTORE_DELETE_FINISHED = exports.ACTION_DATASTORE_DELETE_STARTED = exports.EVENT_ACTION_DATASTORE_DELETE = exports.ACTION_DATASTORE_CREATE_FINISHED = exports.ACTION_DATASTORE_CREATE_STARTED = exports.EVENT_ACTION_DATASTORE_CREATE = exports.EVENT_ACTION_DATASTORE = exports.ACTION_NAVIGATE_FINISHED = exports.ACTION_NAVIGATE_STARTED = exports.EVENT_ACTION_CORE_NAVIGATE = exports.ACTION_STATE_MUTATION_FINISHED = exports.ACTION_STATE_MUTATION_STARTED = exports.EVENT_ACTION_CORE_STATE_MUTATION = exports.EVENT_ACTION_CORE = exports.ACTION_AUTH_SIGNOUT_FINISHED = exports.ACTION_AUTH_SIGNOUT_STARTED = exports.EVENT_ACTION_AUTH_SIGNOUT = exports.EVENT_ACTION_AUTH = exports.STATUS_FINISHED = exports.STATUS_STARTED = exports.ACTION_STATE_MUTATION = exports.ACTION_DATASTORE_UPDATE = exports.ACTION_DATASTORE_DELETE = exports.ACTION_DATASTORE_CREATE = exports.ACTION_NAVIGATE = exports.ACTION_AUTH_SIGNOUT = exports.CATEGORY_CORE = exports.CATEGORY_DATASTORE = exports.CATEGORY_AUTH = exports.UI_EVENT_TYPE_ACTIONS = exports.UI_CHANNEL = void 0;
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
exports.UI_CHANNEL = 'ui';
exports.UI_EVENT_TYPE_ACTIONS = 'actions';
exports.CATEGORY_AUTH = 'auth';
exports.CATEGORY_DATASTORE = 'datastore';
exports.CATEGORY_CORE = 'core';
exports.ACTION_AUTH_SIGNOUT = 'signout';
exports.ACTION_NAVIGATE = 'navigate';
exports.ACTION_DATASTORE_CREATE = 'create';
exports.ACTION_DATASTORE_DELETE = 'delete';
exports.ACTION_DATASTORE_UPDATE = 'update';
exports.ACTION_STATE_MUTATION = 'statemutation';
exports.STATUS_STARTED = 'started';
exports.STATUS_FINISHED = 'finished';
// actions:auth:signout
exports.EVENT_ACTION_AUTH = `${exports.UI_EVENT_TYPE_ACTIONS}:${exports.CATEGORY_AUTH}`;
exports.EVENT_ACTION_AUTH_SIGNOUT = `${exports.EVENT_ACTION_AUTH}:${exports.ACTION_AUTH_SIGNOUT}`;
exports.ACTION_AUTH_SIGNOUT_STARTED = `${exports.EVENT_ACTION_AUTH_SIGNOUT}:${exports.STATUS_STARTED}`;
exports.ACTION_AUTH_SIGNOUT_FINISHED = `${exports.EVENT_ACTION_AUTH_SIGNOUT}:${exports.STATUS_FINISHED}`;
// actions:core
exports.EVENT_ACTION_CORE = `${exports.UI_EVENT_TYPE_ACTIONS}:${exports.CATEGORY_CORE}`;
// actions:core:statemutation
exports.EVENT_ACTION_CORE_STATE_MUTATION = `${exports.EVENT_ACTION_CORE}:${exports.ACTION_STATE_MUTATION}`;
exports.ACTION_STATE_MUTATION_STARTED = `${exports.EVENT_ACTION_CORE_STATE_MUTATION}:${exports.STATUS_STARTED}`;
exports.ACTION_STATE_MUTATION_FINISHED = `${exports.EVENT_ACTION_CORE_STATE_MUTATION}:${exports.STATUS_FINISHED}`;
// actions:core:navigate
exports.EVENT_ACTION_CORE_NAVIGATE = `${exports.EVENT_ACTION_CORE}:${exports.ACTION_NAVIGATE}`;
exports.ACTION_NAVIGATE_STARTED = `${exports.EVENT_ACTION_CORE_NAVIGATE}:${exports.STATUS_STARTED}`;
exports.ACTION_NAVIGATE_FINISHED = `${exports.EVENT_ACTION_CORE_NAVIGATE}:${exports.STATUS_FINISHED}`;
// actions:datastore
exports.EVENT_ACTION_DATASTORE = `${exports.UI_EVENT_TYPE_ACTIONS}:${exports.CATEGORY_DATASTORE}`;
// actions:datastore:create
exports.EVENT_ACTION_DATASTORE_CREATE = `${exports.EVENT_ACTION_DATASTORE}:${exports.ACTION_DATASTORE_CREATE}`;
exports.ACTION_DATASTORE_CREATE_STARTED = `${exports.EVENT_ACTION_DATASTORE_CREATE}:${exports.STATUS_STARTED}`;
exports.ACTION_DATASTORE_CREATE_FINISHED = `${exports.EVENT_ACTION_DATASTORE_CREATE}:${exports.STATUS_FINISHED}`;
// actions:datastore:delete
exports.EVENT_ACTION_DATASTORE_DELETE = `${exports.EVENT_ACTION_DATASTORE}:${exports.ACTION_DATASTORE_DELETE}`;
exports.ACTION_DATASTORE_DELETE_STARTED = `${exports.EVENT_ACTION_DATASTORE_DELETE}:${exports.STATUS_STARTED}`;
exports.ACTION_DATASTORE_DELETE_FINISHED = `${exports.EVENT_ACTION_DATASTORE_DELETE}:${exports.STATUS_FINISHED}`;
// actions:datastore:update
exports.EVENT_ACTION_DATASTORE_UPDATE = `${exports.EVENT_ACTION_DATASTORE}:${exports.ACTION_DATASTORE_UPDATE}`;
exports.ACTION_DATASTORE_UPDATE_STARTED = `${exports.EVENT_ACTION_DATASTORE_UPDATE}:${exports.STATUS_STARTED}`;
exports.ACTION_DATASTORE_UPDATE_FINISHED = `${exports.EVENT_ACTION_DATASTORE_UPDATE}:${exports.STATUS_FINISHED}`;
exports.DATASTORE_QUERY_BY_ID_ERROR = 'Error querying datastore item by id';
exports.constantsString = `export const UI_CHANNEL = 'ui';
export const UI_EVENT_TYPE_ACTIONS = 'actions';
export const CATEGORY_AUTH = 'auth';
export const CATEGORY_DATASTORE = 'datastore';
export const CATEGORY_CORE = 'core';
export const ACTION_AUTH_SIGNOUT = 'signout';
export const ACTION_NAVIGATE = 'navigate';
export const ACTION_DATASTORE_CREATE = 'create';
export const ACTION_DATASTORE_DELETE = 'delete';
export const ACTION_DATASTORE_UPDATE = 'update';
export const ACTION_STATE_MUTATION = 'statemutation';
export const STATUS_STARTED = 'started';
export const STATUS_FINISHED = 'finished';
export const EVENT_ACTION_AUTH = \`\${UI_EVENT_TYPE_ACTIONS}:\${CATEGORY_AUTH}\`;
export const EVENT_ACTION_AUTH_SIGNOUT = \`\${EVENT_ACTION_AUTH}:\${ACTION_AUTH_SIGNOUT}\`;
export const ACTION_AUTH_SIGNOUT_STARTED = \`\${EVENT_ACTION_AUTH_SIGNOUT}:\${STATUS_STARTED}\`;
export const ACTION_AUTH_SIGNOUT_FINISHED = \`\${EVENT_ACTION_AUTH_SIGNOUT}:\${STATUS_FINISHED}\`;
export const EVENT_ACTION_CORE = \`\${UI_EVENT_TYPE_ACTIONS}:\${CATEGORY_CORE}\`;
export const EVENT_ACTION_CORE_STATE_MUTATION = \`\${EVENT_ACTION_CORE}:\${ACTION_STATE_MUTATION}\`;
export const ACTION_STATE_MUTATION_STARTED = \`\${EVENT_ACTION_CORE_STATE_MUTATION}:\${STATUS_STARTED}\`;
export const ACTION_STATE_MUTATION_FINISHED = \`\${EVENT_ACTION_CORE_STATE_MUTATION}:\${STATUS_FINISHED}\`;
export const EVENT_ACTION_CORE_NAVIGATE = \`\${EVENT_ACTION_CORE}:\${ACTION_NAVIGATE}\`;
export const ACTION_NAVIGATE_STARTED = \`\${EVENT_ACTION_CORE_NAVIGATE}:\${STATUS_STARTED}\`;
export const ACTION_NAVIGATE_FINISHED = \`\${EVENT_ACTION_CORE_NAVIGATE}:\${STATUS_FINISHED}\`;
export const EVENT_ACTION_DATASTORE = \`\${UI_EVENT_TYPE_ACTIONS}:\${CATEGORY_DATASTORE}\`;
export const EVENT_ACTION_DATASTORE_CREATE = \`\${EVENT_ACTION_DATASTORE}:\${ACTION_DATASTORE_CREATE}\`;
export const ACTION_DATASTORE_CREATE_STARTED = \`\${EVENT_ACTION_DATASTORE_CREATE}:\${STATUS_STARTED}\`;
export const ACTION_DATASTORE_CREATE_FINISHED = \`\${EVENT_ACTION_DATASTORE_CREATE}:\${STATUS_FINISHED}\`;
export const EVENT_ACTION_DATASTORE_DELETE = \`\${EVENT_ACTION_DATASTORE}:\${ACTION_DATASTORE_DELETE}\`;
export const ACTION_DATASTORE_DELETE_STARTED = \`\${EVENT_ACTION_DATASTORE_DELETE}:\${STATUS_STARTED}\`;
export const ACTION_DATASTORE_DELETE_FINISHED = \`\${EVENT_ACTION_DATASTORE_DELETE}:\${STATUS_FINISHED}\`;
export const EVENT_ACTION_DATASTORE_UPDATE = \`\${EVENT_ACTION_DATASTORE}:\${ACTION_DATASTORE_UPDATE}\`;
export const ACTION_DATASTORE_UPDATE_STARTED = \`\${EVENT_ACTION_DATASTORE_UPDATE}:\${STATUS_STARTED}\`;
export const ACTION_DATASTORE_UPDATE_FINISHED = \`\${EVENT_ACTION_DATASTORE_UPDATE}:\${STATUS_FINISHED}\`;
export const DATASTORE_QUERY_BY_ID_ERROR = 'Error querying datastore item by id';`;
