// @aws-amplify/graphql-api-construct/lib/internal/asset-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetProvider = void 0;
const fs = require("fs");
const path = require("path");
const cdk = require("aws-cdk-lib");
const aws_s3_assets_1 = require("aws-cdk-lib/aws-s3-assets");
const TEMP_PREFIX = 'transformer-assets';
const FUNCTION_PREFIX = 'functions';
const RESOLVER_PREFIX = 'resolvers';
/**
 * The asset provider bridges the gap between creation of file assets in the transformer (which provide a name+contents tuple)
 * with the path method which is used in CDK.
 * The CDK S3 asset require the contents to be written to the file system first.
 * The asset provider writes to a temporary directory before creating the CDK S3 asset.
 *
 */
class AssetProvider {
    constructor(scope) {
        this.resolverAssets = {};
        this.tempAssetDir = cdk.FileSystem.mkdtemp(`${TEMP_PREFIX}-${scope.node.addr}`);
    }
    /**
     * Creates a new CDK S3 asset. The file contents in assetProps is first stored in a temporary file that is referenced by the CDK S3 asset.
     * @param assetScope the parent of the asset
     * @param assetId unique ID for CDK S3 asset
     * @param assetProps name and contents of file to be added to CDK S3 asset
     * @returns the CDK S3 asset
     */
    provide(assetScope, assetId, assetProps) {
        return new aws_s3_assets_1.Asset(assetScope, assetId, { path: this.addAsset(assetProps.fileName, assetProps.fileContent) });
    }
    addAsset(fileName, contents) {
        this.trackResolverAsset(fileName, contents);
        const writableContents = this.isContentsAReference(fileName) ? this.dereferenceContents(contents) : contents;
        const filePath = path.join(this.tempAssetDir, fileName);
        const fileDirName = path.dirname(filePath);
        if (!fs.existsSync(fileDirName)) {
            fs.mkdirSync(fileDirName, { recursive: true });
        }
        fs.writeFileSync(filePath, writableContents);
        return filePath;
    }
    isContentsAReference(fileName) {
        return fileName.startsWith(FUNCTION_PREFIX);
    }
    dereferenceContents(contents) {
        return fs.readFileSync(contents);
    }
    trackResolverAsset(fileName, contents) {
        if (fileName.startsWith(RESOLVER_PREFIX)) {
            const resolverFileName = fileName.split('/')[1];
            this.resolverAssets[resolverFileName] = contents;
        }
    }
}
exports.AssetProvider = AssetProvider;
