// @aws-amplify/graphql-api-construct/lib/internal/codegen-assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodegenAssets = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_s3_1 = require("aws-cdk-lib/aws-s3");
const aws_s3_deployment_1 = require("aws-cdk-lib/aws-s3-deployment");
const constructs_1 = require("constructs");
const MODEL_SCHEMA_KEY = 'model-schema.graphql';
const CONSOLE_SERVICE_ENDPOINT = aws_cdk_lib_1.Fn.join('', ['https://', aws_cdk_lib_1.Fn.ref('AWS::Region'), '.console.aws.amazon.com/amplify']);
/**
 * Construct an S3 URI string for a given bucket and key.
 * @param bucket the bucket to embed in the uri string
 * @param key the key to embed in the uri string
 * @returns the uri string representation.
 */
const getS3UriForBucketAndKey = (bucket, key) => `s3://${bucket.bucketName}/${key}`;
/**
 * Construct which creates a bucket, and uploads file assets required for codegen to run.
 * Pointers to these resources are persisted in the stack outputs.
 */
class CodegenAssets extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const bucket = new aws_s3_1.Bucket(this, `${id}Bucket`, {
            removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
            autoDeleteObjects: true,
            // Enabling CORS to allow console to access the codegen assets.
            cors: [
                {
                    allowedMethods: [aws_s3_1.HttpMethods.GET, aws_s3_1.HttpMethods.HEAD],
                    allowedHeaders: ['*'],
                    allowedOrigins: [CONSOLE_SERVICE_ENDPOINT],
                },
            ],
        });
        const deployment = new aws_s3_deployment_1.BucketDeployment(this, `${id}Deployment`, {
            destinationBucket: bucket,
            sources: [aws_s3_deployment_1.Source.data(MODEL_SCHEMA_KEY, props.modelSchema)],
            // Bucket deployment uses a Lambda that runs AWS S3 CLI to transfer assets to destination bucket.
            // That Lambda requires higher memory setting to run fast even when processing small assets (less than 1kB).
            // This setting has been established experimentally. Benchmark can be found in pull request description that established it.
            // The value has been chosen to prefer the lowest cost (run time * memory demand) while providing reasonable performance.
            memoryLimit: 1536,
        });
        this.modelSchemaS3Uri = getS3UriForBucketAndKey(deployment.deployedBucket, MODEL_SCHEMA_KEY);
    }
}
exports.CodegenAssets = CodegenAssets;
