// @aws-amplify/graphql-api-construct/lib/internal/conflict-resolution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToResolverConfig = void 0;
/**
 * Convert project conflict resolution config to transformer ResolverConfig object.
 * @param param0 input shape
 * @param param0.project the project level config
 * @param param0.models the model-specific override config
 * @returns the transformer representation
 */
const convertToResolverConfig = ({ project, models }) => ({
    project: project && convertToSyncConfig(project),
    models: models && Object.fromEntries(Object.entries(models).map(([modelName, strategy]) => [modelName, convertToSyncConfig(strategy)])),
});
exports.convertToResolverConfig = convertToResolverConfig;
/**
 * Convert from external to internal representation of conflict resolution config.
 * @param strategy the strategy to convert
 * @returns the converted strategy
 */
const convertToSyncConfig = (strategy) => {
    switch (strategy.handlerType) {
        case 'OPTIMISTIC_CONCURRENCY':
            return {
                ConflictHandler: "OPTIMISTIC_CONCURRENCY" /* ConflictHandlerType.OPTIMISTIC */,
                ConflictDetection: strategy.detectionType,
            };
        case 'AUTOMERGE':
            return {
                ConflictHandler: "AUTOMERGE" /* ConflictHandlerType.AUTOMERGE */,
                ConflictDetection: strategy.detectionType,
            };
        case 'LAMBDA':
            return {
                ConflictHandler: "LAMBDA" /* ConflictHandlerType.LAMBDA */,
                ConflictDetection: strategy.detectionType,
                LambdaConflictHandler: {
                    name: strategy.conflictHandler.functionName,
                    lambdaArn: strategy.conflictHandler.functionArn,
                },
            };
        default:
            throw new Error(`Encountered unexpected handlerType: ${strategy.handlerType}`);
    }
};
