// @aws-amplify/codegen-ui-react/dist/lib/imports/import-mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportMapping = exports.ImportValue = exports.ImportSource = void 0;
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
var ImportSource;
(function (ImportSource) {
    ImportSource["REACT"] = "react";
    ImportSource["UI_REACT"] = "@aws-amplify/ui-react";
    ImportSource["UI_REACT_INTERNAL"] = "@aws-amplify/ui-react/internal";
    ImportSource["AMPLIFY_DATASTORE"] = "@aws-amplify/datastore";
    ImportSource["REACT_STORAGE"] = "@aws-amplify/ui-react-storage";
    ImportSource["LOCAL_MODELS"] = "../models";
    ImportSource["LOCAL_SCHEMA"] = "../models/schema";
    ImportSource["UTILS"] = "./utils";
    ImportSource["AMPLIFY"] = "aws-amplify";
    ImportSource["AMPLIFY_API"] = "aws-amplify/api";
    ImportSource["AMPLIFY_AUTH"] = "aws-amplify/auth";
    ImportSource["AMPLIFY_UTILS"] = "aws-amplify/utils";
    ImportSource["AMPLIFY_DATASTORE_V6"] = "aws-amplify/datastore";
})(ImportSource = exports.ImportSource || (exports.ImportSource = {}));
var ImportValue;
(function (ImportValue) {
    ImportValue["GET_OVERRIDE_PROPS"] = "getOverrideProps";
    ImportValue["USE_AUTH"] = "useAuth";
    ImportValue["AUTH"] = "Auth";
    ImportValue["SIGN_OUT"] = "signOut";
    ImportValue["FETCH_USER_ATTRIBUTES"] = "fetchUserAttributes";
    ImportValue["GET_OVERRIDES_FROM_VARIANTS"] = "getOverridesFromVariants";
    ImportValue["USE_BREAKPOINT_VALUE"] = "useBreakpointValue";
    ImportValue["VARIANT"] = "Variant";
    ImportValue["SORT_DIRECTION"] = "SortDirection";
    ImportValue["SORT_PREDICATE"] = "SortPredicate";
    ImportValue["CREATE_DATA_STORE_PREDICATE"] = "createDataStorePredicate";
    ImportValue["USE_DATA_STORE_BINDING"] = "useDataStoreBinding";
    ImportValue["CREATE_THEME"] = "createTheme";
    ImportValue["MERGE_VARIANTS_OVERRIDES"] = "mergeVariantsAndOverrides";
    ImportValue["USE_NAVIGATE_ACTION"] = "useNavigateAction";
    ImportValue["USE_DATA_STORE_CREATE_ACTION"] = "useDataStoreCreateAction";
    ImportValue["USE_DATA_STORE_UPDATE_ACTION"] = "useDataStoreUpdateAction";
    ImportValue["USE_DATA_STORE_DELETE_ACTION"] = "useDataStoreDeleteAction";
    ImportValue["USE_AUTH_SIGN_OUT_ACTION"] = "useAuthSignOutAction";
    ImportValue["USE_STATE_MUTATION_ACTION"] = "useStateMutationAction";
    ImportValue["USE_EFFECT"] = "useEffect";
    ImportValue["VALIDATE_FIELD"] = "validateField";
    ImportValue["VALIDATE_FIELD_CODEGEN"] = "validateField";
    ImportValue["FORMATTER"] = "formatter";
    ImportValue["FETCH_BY_PATH"] = "fetchByPath";
    ImportValue["PROCESS_FILE"] = "processFile";
    ImportValue["DEFAULT_THEME"] = "defaultTheme";
    ImportValue["USE_STATE"] = "useState";
    ImportValue["API"] = "API";
    ImportValue["HUB"] = "Hub";
    ImportValue["DATASTORE"] = "DataStore";
    ImportValue["PAGINATION"] = "Pagination";
    ImportValue["PLACEHOLDER"] = "Placeholder";
    ImportValue["GENERATE_CLIENT"] = "generateClient";
})(ImportValue = exports.ImportValue || (exports.ImportValue = {}));
exports.ImportMapping = {
    [ImportValue.API]: ImportSource.AMPLIFY,
    [ImportValue.AUTH]: ImportSource.AMPLIFY,
    [ImportValue.SIGN_OUT]: ImportSource.AMPLIFY_AUTH,
    [ImportValue.HUB]: ImportSource.AMPLIFY,
    [ImportValue.DATASTORE]: ImportSource.AMPLIFY,
    [ImportValue.GENERATE_CLIENT]: ImportSource.AMPLIFY_API,
    [ImportValue.SORT_DIRECTION]: ImportSource.AMPLIFY_DATASTORE,
    [ImportValue.SORT_PREDICATE]: ImportSource.AMPLIFY_DATASTORE,
    [ImportValue.CREATE_THEME]: ImportSource.UI_REACT,
    [ImportValue.DEFAULT_THEME]: ImportSource.UI_REACT,
    [ImportValue.PAGINATION]: ImportSource.UI_REACT,
    [ImportValue.PLACEHOLDER]: ImportSource.UI_REACT,
    [ImportValue.USE_BREAKPOINT_VALUE]: ImportSource.UI_REACT,
    [ImportValue.USE_AUTH]: ImportSource.UI_REACT_INTERNAL,
    [ImportValue.CREATE_DATA_STORE_PREDICATE]: ImportSource.UTILS,
    [ImportValue.USE_DATA_STORE_BINDING]: ImportSource.UTILS,
    [ImportValue.USE_NAVIGATE_ACTION]: ImportSource.UTILS,
    [ImportValue.USE_DATA_STORE_CREATE_ACTION]: ImportSource.UTILS,
    [ImportValue.USE_DATA_STORE_UPDATE_ACTION]: ImportSource.UTILS,
    [ImportValue.USE_DATA_STORE_DELETE_ACTION]: ImportSource.UTILS,
    [ImportValue.USE_AUTH_SIGN_OUT_ACTION]: ImportSource.UTILS,
    [ImportValue.USE_STATE_MUTATION_ACTION]: ImportSource.UTILS,
    [ImportValue.GET_OVERRIDE_PROPS]: ImportSource.UTILS,
    [ImportValue.MERGE_VARIANTS_OVERRIDES]: ImportSource.UTILS,
    [ImportValue.GET_OVERRIDES_FROM_VARIANTS]: ImportSource.UTILS,
    [ImportValue.VARIANT]: ImportSource.UTILS,
    [ImportValue.FORMATTER]: ImportSource.UTILS,
    [ImportValue.VALIDATE_FIELD]: ImportSource.UTILS,
    [ImportValue.FETCH_BY_PATH]: ImportSource.UTILS,
    [ImportValue.PROCESS_FILE]: ImportSource.UTILS,
    [ImportValue.USE_EFFECT]: ImportSource.REACT,
    [ImportValue.USE_STATE]: ImportSource.REACT,
    [ImportValue.FETCH_USER_ATTRIBUTES]: ImportSource.AMPLIFY_AUTH,
};
