// @aws-amplify/deployed-backend-client/lib/backend_output_fetcher_factory.js
import { AppNameAndBranchMainStackNameResolver, } from './stack-name-resolvers/app_name_and_branch_main_stack_name_resolver.js';
import { PassThroughMainStackNameResolver, } from './stack-name-resolvers/passthrough_main_stack_name_resolver.js';
import { BackendIdentifierMainStackNameResolver } from './stack-name-resolvers/backend_identifier_main_stack_name_resolver.js';
import { StackMetadataBackendOutputRetrievalStrategy } from './stack_metadata_output_retrieval_strategy.js';
/**
 * Asserts that a BackendIdentifier is a BackendIdentifier
 */
export const isBackendIdentifier = (backendIdentifier) => {
    return ('namespace' in backendIdentifier &&
        'name' in backendIdentifier &&
        'type' in backendIdentifier);
};
/**
 * Asserts that a BackendIdentifier is a AppNameAndBranchBackendIdentifier
 */
export const isAppNameAndBranchIdentifier = (backendIdentifier) => {
    return 'appName' in backendIdentifier;
};
/**
 * Asserts that a BackendIdentifier is a StackIdentifier
 */
export const isStackIdentifier = (backendIdentifier) => {
    return 'stackName' in backendIdentifier;
};
/**
 * Constructs an OutputRetrievalStrategy depending on the type of StackIdentifier
 */
export class BackendOutputFetcherFactory {
    cfnClient;
    amplifyClient;
    /**
     * Instantiates the factory
     */
    constructor(cfnClient, amplifyClient) {
        this.cfnClient = cfnClient;
        this.amplifyClient = amplifyClient;
    }
    getStrategy = (backendIdentifier) => {
        if (isStackIdentifier(backendIdentifier)) {
            return new StackMetadataBackendOutputRetrievalStrategy(this.cfnClient, new PassThroughMainStackNameResolver(backendIdentifier));
        }
        else if (isBackendIdentifier(backendIdentifier)) {
            return new StackMetadataBackendOutputRetrievalStrategy(this.cfnClient, new BackendIdentifierMainStackNameResolver(backendIdentifier));
        }
        return new StackMetadataBackendOutputRetrievalStrategy(this.cfnClient, new AppNameAndBranchMainStackNameResolver(this.amplifyClient, backendIdentifier));
    };
}
