// @aws-amplify/deployed-backend-client/lib/backend_output_client_factory.js
import { DefaultBackendOutputClient } from './backend_output_client.js';
import { CloudFormationClient } from '@aws-sdk/client-cloudformation';
import { AmplifyClient } from '@aws-sdk/client-amplify';
export var BackendOutputClientErrorType;
(function (BackendOutputClientErrorType) {
    BackendOutputClientErrorType["METADATA_RETRIEVAL_ERROR"] = "MetadataRetrievalError";
    BackendOutputClientErrorType["NO_OUTPUTS_FOUND"] = "NoOutputsFound";
    BackendOutputClientErrorType["DEPLOYMENT_IN_PROGRESS"] = "DeploymentInProgress";
    BackendOutputClientErrorType["NO_STACK_FOUND"] = "NoStackFound";
    BackendOutputClientErrorType["CREDENTIALS_ERROR"] = "CredentialsError";
    BackendOutputClientErrorType["ACCESS_DENIED"] = "AccessDenied";
})(BackendOutputClientErrorType || (BackendOutputClientErrorType = {}));
/**
 * Error type for BackendOutputClientError
 */
export class BackendOutputClientError extends Error {
    code;
    /**
     * Constructor for BackendOutputClientError
     */
    constructor(code, message, options) {
        super(message, options);
        this.code = code;
    }
}
/**
 * Factory to create a backend metadata reader
 */
export class BackendOutputClientFactory {
    /**
     * Returns a single instance of BackendOutputClient
     */
    static getInstance = (awsClientProvider) => {
        return new DefaultBackendOutputClient(awsClientProvider?.getCloudFormationClient() ??
            new CloudFormationClient(), awsClientProvider?.getAmplifyClient() ?? new AmplifyClient());
    };
}
