// @aws-amplify/deployed-backend-client/lib/deployed_backend_client_factory.js
import { DefaultDeployedBackendClient } from './deployed_backend_client.js';
import { BackendOutputClientFactory } from './backend_output_client_factory.js';
import { DeployedResourcesEnumerator } from './deployed-backend-client/deployed_resources_enumerator.js';
import { StackStatusMapper } from './deployed-backend-client/stack_status_mapper.js';
import { ArnGenerator } from './deployed-backend-client/arn_generator.js';
import { ArnParser } from './deployed-backend-client/arn_parser.js';
export var ConflictResolutionMode;
(function (ConflictResolutionMode) {
    ConflictResolutionMode["LAMBDA"] = "LAMBDA";
    ConflictResolutionMode["OPTIMISTIC_CONCURRENCY"] = "OPTIMISTIC_CONCURRENCY";
    ConflictResolutionMode["AUTOMERGE"] = "AUTOMERGE";
})(ConflictResolutionMode || (ConflictResolutionMode = {}));
export var ApiAuthType;
(function (ApiAuthType) {
    ApiAuthType["API_KEY"] = "API_KEY";
    ApiAuthType["AWS_LAMBDA"] = "AWS_LAMBDA";
    ApiAuthType["AWS_IAM"] = "AWS_IAM";
    ApiAuthType["OPENID_CONNECT"] = "OPENID_CONNECT";
    ApiAuthType["AMAZON_COGNITO_USER_POOLS"] = "AMAZON_COGNITO_USER_POOLS";
})(ApiAuthType || (ApiAuthType = {}));
export var BackendDeploymentStatus;
(function (BackendDeploymentStatus) {
    BackendDeploymentStatus["DEPLOYED"] = "DEPLOYED";
    BackendDeploymentStatus["FAILED"] = "FAILED";
    BackendDeploymentStatus["DEPLOYING"] = "DEPLOYING";
    BackendDeploymentStatus["DELETING"] = "DELETING";
    BackendDeploymentStatus["DELETED"] = "DELETED";
    BackendDeploymentStatus["UNKNOWN"] = "UNKNOWN";
})(BackendDeploymentStatus || (BackendDeploymentStatus = {}));
export var BackendStatus;
(function (BackendStatus) {
    BackendStatus["DELETE_FAILED"] = "DELETE_FAILED";
})(BackendStatus || (BackendStatus = {}));
/**
 * Factory to create a DeploymentClient
 */
export class DeployedBackendClientFactory {
    /**
     * Returns a single instance of DeploymentClient
     */
    getInstance(awsClientProvider) {
        const stackStatusMapper = new StackStatusMapper();
        const arnGenerator = new ArnGenerator();
        const arnParser = new ArnParser();
        const deployedResourcesEnumerator = new DeployedResourcesEnumerator(stackStatusMapper, arnGenerator, arnParser);
        return new DefaultDeployedBackendClient(awsClientProvider.getCloudFormationClient(), awsClientProvider.getS3Client(), BackendOutputClientFactory.getInstance(awsClientProvider), deployedResourcesEnumerator, stackStatusMapper, arnParser);
    }
}
