// @aws-amplify/graphql-auth-transformer/lib/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SEARCHABLE_AGGREGATE_TYPES = exports.RELATIONAL_DIRECTIVES = exports.NONE_DS = exports.DENIED_FIELDS = exports.NULL_ALLOWED_FIELDS = exports.ALLOWED_FIELDS = exports.IS_AUTHORIZED_FLAG = exports.API_KEY_AUTH_TYPE = exports.LAMBDA_AUTH_TYPE = exports.IAM_AUTH_TYPE = exports.OIDC_AUTH_TYPE = exports.COGNITO_AUTH_TYPE = exports.AUTH_PROVIDER_DIRECTIVE_MAP = exports.MODEL_OPERATIONS = exports.READ_MODEL_OPERATIONS = exports.AUTH_NON_MODEL_TYPES = exports.ON_DELETE_FIELD = exports.ON_UPDATE_FIELD = exports.ON_CREATE_FIELD = exports.IDENTITY_CLAIM_DELIMITER = exports.DEFAULT_GROUP_CLAIM = exports.DEFAULT_COGNITO_IDENTITY_CLAIM = exports.DEFAULT_UNIQUE_IDENTITY_CLAIM = exports.DEFAULT_IDENTITY_CLAIM = exports.DEFAULT_GROUPS_FIELD = exports.DEFAULT_OWNER_FIELD = void 0;
exports.DEFAULT_OWNER_FIELD = 'owner';
exports.DEFAULT_GROUPS_FIELD = 'groups';
exports.DEFAULT_IDENTITY_CLAIM = 'username';
exports.DEFAULT_UNIQUE_IDENTITY_CLAIM = 'sub::username';
exports.DEFAULT_COGNITO_IDENTITY_CLAIM = 'cognito:username';
exports.DEFAULT_GROUP_CLAIM = 'cognito:groups';
exports.IDENTITY_CLAIM_DELIMITER = '::';
exports.ON_CREATE_FIELD = 'onCreate';
exports.ON_UPDATE_FIELD = 'onUpdate';
exports.ON_DELETE_FIELD = 'onDelete';
exports.AUTH_NON_MODEL_TYPES = 'authNonModelTypes';
exports.READ_MODEL_OPERATIONS = ['get', 'list', 'search', 'listen', 'sync'];
exports.MODEL_OPERATIONS = ['create', 'update', 'delete', ...exports.READ_MODEL_OPERATIONS];
exports.AUTH_PROVIDER_DIRECTIVE_MAP = new Map([
    ['apiKey', 'aws_api_key'],
    ['iam', 'aws_iam'],
    ['identityPool', 'aws_iam'],
    ['oidc', 'aws_oidc'],
    ['userPools', 'aws_cognito_user_pools'],
    ['function', 'aws_lambda'],
]);
exports.COGNITO_AUTH_TYPE = 'User Pool Authorization';
exports.OIDC_AUTH_TYPE = 'Open ID Connect Authorization';
exports.IAM_AUTH_TYPE = 'IAM Authorization';
exports.LAMBDA_AUTH_TYPE = 'Lambda Authorization';
exports.API_KEY_AUTH_TYPE = 'API Key Authorization';
exports.IS_AUTHORIZED_FLAG = 'isAuthorized';
exports.ALLOWED_FIELDS = 'allowedFields';
exports.NULL_ALLOWED_FIELDS = 'nullAllowedFields';
exports.DENIED_FIELDS = 'deniedFields';
exports.NONE_DS = 'NONE_DS';
exports.RELATIONAL_DIRECTIVES = ['hasOne', 'belongsTo', 'hasMany', 'manyToMany'];
exports.SEARCHABLE_AGGREGATE_TYPES = [
    'SearchableAggregateResult',
    'SearchableAggregateScalarResult',
    'SearchableAggregateBucketResult',
    'SearchableAggregateBucketResultItem',
];
