// @aws-amplify/graphql-relational-transformer/lib/has-one/has-one-directive-ddb-fields-transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HasOneDirectiveDDBFieldsTransformer = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const utils_1 = require("../utils");
const generator_factory_1 = require("../resolver/generator-factory");
class HasOneDirectiveDDBFieldsTransformer {
    constructor() {
        this.dbType = graphql_transformer_core_1.DDB_DB_TYPE;
        this.prepare = (context, config) => {
            const modelName = config.object.name.value;
            (0, utils_1.registerHasOneForeignKeyMappings)({
                transformParameters: context.transformParameters,
                resourceHelper: context.resourceHelper,
                thisTypeName: modelName,
                thisFieldName: config.field.name.value,
                relatedType: config.relatedType,
            });
        };
        this.transformSchema = (context, config) => {
            config.relatedTypeIndex = (0, utils_1.getRelatedTypeIndex)(config, context);
        };
        this.generateResolvers = (context, config) => {
            const generator = (0, generator_factory_1.getGenerator)(this.dbType);
            generator.makeHasOneGetItemConnectionWithKeyResolver(config, context);
        };
        this.validate = (context, config) => {
            (0, utils_1.ensureFieldsArray)(config);
            config.fieldNodes = (0, utils_1.getFieldsNodes)(config, context);
        };
    }
}
exports.HasOneDirectiveDDBFieldsTransformer = HasOneDirectiveDDBFieldsTransformer;
