// @aws-amplify/graphql-docs-generator/lib/generator/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExternalFragmentPartial = exports.getOperationPartial = exports.getTemplatePartials = exports.buildSchema = exports.lowerCaseFirstLetter = exports.generateQueries = exports.generateSubscriptions = exports.generateMutations = void 0;
__exportStar(require("./types"), exports);
const generate_1 = require("./generate");
var generateAllOperations_1 = require("./generateAllOperations");
Object.defineProperty(exports, "generateMutations", { enumerable: true, get: function () { return generateAllOperations_1.generateMutations; } });
Object.defineProperty(exports, "generateSubscriptions", { enumerable: true, get: function () { return generateAllOperations_1.generateSubscriptions; } });
Object.defineProperty(exports, "generateQueries", { enumerable: true, get: function () { return generateAllOperations_1.generateQueries; } });
Object.defineProperty(exports, "lowerCaseFirstLetter", { enumerable: true, get: function () { return generateAllOperations_1.lowerCaseFirstLetter; } });
var loading_1 = require("./utils/loading");
Object.defineProperty(exports, "buildSchema", { enumerable: true, get: function () { return loading_1.buildSchema; } });
var templates_1 = require("./utils/templates");
Object.defineProperty(exports, "getTemplatePartials", { enumerable: true, get: function () { return templates_1.getTemplatePartials; } });
Object.defineProperty(exports, "getOperationPartial", { enumerable: true, get: function () { return templates_1.getOperationPartial; } });
Object.defineProperty(exports, "getExternalFragmentPartial", { enumerable: true, get: function () { return templates_1.getExternalFragmentPartial; } });
exports.default = generate_1.default;
