// @aws-amplify/codegen-ui-react/dist/lib/forms/form-renderer-helper/value-props.js
"use strict";
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetFunctionCheck = exports.renderValueAttribute = exports.renderDefaultValueAttribute = exports.getConvertedValueAttribute = exports.isControlledComponent = exports.ControlledComponents = void 0;
const typescript_1 = require("typescript");
const form_state_1 = require("./form-state");
const render_checkers_1 = require("./render-checkers");
const typescript_type_map_1 = require("./typescript-type-map");
exports.ControlledComponents = [
    'StepperField',
    'SliderField',
    'SelectField',
    'ToggleButton',
    'SwitchField',
    'TextField',
];
/**
 * given the component returns true if the component is a controlled component
 *
 * @param componentType
 * @returns
 */
const isControlledComponent = (componentType) => exports.ControlledComponents.includes(componentType);
exports.isControlledComponent = isControlledComponent;
const getConvertedValueAttribute = (dataType, studioFormComponentType) => {
    switch (dataType) {
        case 'AWSDateTime':
            return (valueIdentifier) => typescript_1.factory.createBinaryExpression(valueIdentifier, typescript_1.factory.createToken(typescript_1.SyntaxKind.AmpersandAmpersandToken), typescript_1.factory.createCallExpression(typescript_1.factory.createIdentifier('convertToLocal'), undefined, [
                typescript_1.factory.createNewExpression(typescript_1.factory.createIdentifier('Date'), undefined, [valueIdentifier]),
            ]));
        case 'AWSTimestamp':
            if (studioFormComponentType === 'DateTimeField') {
                return (valueIdentifier) => typescript_1.factory.createBinaryExpression(valueIdentifier, typescript_1.factory.createToken(typescript_1.SyntaxKind.AmpersandAmpersandToken), typescript_1.factory.createCallExpression(typescript_1.factory.createIdentifier('convertToLocal'), undefined, [
                    typescript_1.factory.createCallExpression(typescript_1.factory.createIdentifier('convertTimeStampToDate'), undefined, [
                        valueIdentifier,
                    ]),
                ]));
            }
            break;
        default:
            return undefined;
    }
    return undefined;
};
exports.getConvertedValueAttribute = getConvertedValueAttribute;
/**
 *
 * @param stateName name of the state name to render for the default value
 * @param { dataType } the dataType
 * @returns
 */
const renderDefaultValueAttribute = (stateName, { dataType, studioFormComponentType }, componentType) => {
    const identifier = typescript_1.factory.createIdentifier(stateName);
    let expression = typescript_1.factory.createJsxExpression(undefined, identifier);
    if (dataType && typeof dataType !== 'object') {
        const convertedValueAttribute = (0, exports.getConvertedValueAttribute)(dataType, studioFormComponentType || componentType);
        if (convertedValueAttribute) {
            expression = typescript_1.factory.createJsxExpression(undefined, convertedValueAttribute(identifier));
        }
    }
    return typescript_1.factory.createJsxAttribute(componentType === 'TextField' || componentType === 'TextAreaField'
        ? typescript_1.factory.createIdentifier('value')
        : typescript_1.factory.createIdentifier('defaultValue'), expression);
};
exports.renderDefaultValueAttribute = renderDefaultValueAttribute;
const renderValueAttribute = ({ fieldConfig, componentName, currentValueIdentifier, }) => {
    var _a;
    const componentType = (_a = fieldConfig.studioFormComponentType) !== null && _a !== void 0 ? _a : fieldConfig.componentType;
    const shouldGetForUncontrolled = (0, render_checkers_1.shouldWrapInArrayField)(fieldConfig);
    const { dataType, studioFormComponentType } = fieldConfig;
    let valueIdentifier = currentValueIdentifier || getValueIdentifier(componentName, componentType);
    if ((0, render_checkers_1.shouldWrapInArrayField)(fieldConfig)) {
        if (fieldConfig.relationship) {
            valueIdentifier = typescript_1.factory.createIdentifier((0, form_state_1.getCurrentDisplayValueName)(componentName));
        }
        else {
            valueIdentifier = typescript_1.factory.createIdentifier((0, form_state_1.getCurrentValueName)(componentName));
        }
    }
    let renderedFieldName = fieldConfig.sanitizedFieldName || componentName;
    if (fieldConfig.isArray) {
        renderedFieldName = (0, form_state_1.getCurrentValueName)(renderedFieldName);
    }
    let fieldNameIdentifier = typescript_1.factory.createIdentifier(renderedFieldName);
    // If a component has a '.', it's nested and needs an object ref.
    // but if it's an array, it needs to use the currentArrayValue whether it's nested or not
    // The ArrayField component will update the nested value.
    if (componentName.includes('.') && !fieldConfig.isArray) {
        const [parent, child] = componentName.split('.');
        // if it's nested default to an empty string because initial value is undefined.
        // favoriteThings["animals"] ?? ""
        fieldNameIdentifier = typescript_1.factory.createBinaryExpression(typescript_1.factory.createElementAccessExpression(typescript_1.factory.createIdentifier(parent), typescript_1.factory.createStringLiteral(child)), typescript_1.factory.createToken(typescript_1.SyntaxKind.QuestionQuestionToken), typescript_1.factory.createStringLiteral(''));
    }
    let controlledExpression = typescript_1.factory.createJsxExpression(undefined, fieldNameIdentifier);
    if (dataType && typeof dataType !== 'object') {
        const convertedValueAttribute = (0, exports.getConvertedValueAttribute)(dataType, studioFormComponentType || componentType);
        if (convertedValueAttribute) {
            controlledExpression = typescript_1.factory.createJsxExpression(undefined, convertedValueAttribute(valueIdentifier));
        }
    }
    const controlledComponentToAttributesMap = {
        ToggleButton: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('isPressed'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        SliderField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        SelectField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        StepperField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        SwitchField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('isChecked'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        CheckboxField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('checked'), typescript_1.factory.createJsxExpression(undefined, valueIdentifier)),
        TextField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        DateTimeField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        IPAddressField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        DateField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        TimeField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        NumberField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        URLField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        PhoneNumberField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
        EmailField: typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), controlledExpression),
    };
    if (controlledComponentToAttributesMap[componentType]) {
        return controlledComponentToAttributesMap[componentType];
    }
    // TODO: all components should be controlled once conversions are solid
    if (shouldGetForUncontrolled) {
        let expression = typescript_1.factory.createJsxExpression(undefined, valueIdentifier);
        if (dataType && typeof dataType !== 'object') {
            const convertedValueAttribute = (0, exports.getConvertedValueAttribute)(dataType, studioFormComponentType || componentType);
            if (convertedValueAttribute) {
                expression = typescript_1.factory.createJsxExpression(undefined, convertedValueAttribute(valueIdentifier));
            }
        }
        return typescript_1.factory.createJsxAttribute(typescript_1.factory.createIdentifier('value'), expression);
    }
    return undefined;
};
exports.renderValueAttribute = renderValueAttribute;
/**
 * returns resetFunctionCheck Expression if a create form is datastore enabled
 *
 * @param studioForm {StudioForm}
 * @returns
 */
const resetFunctionCheck = ({ dataSourceType, formActionType, }) => {
    if (dataSourceType === 'DataStore' && formActionType === 'create') {
        return [
            typescript_1.factory.createIfStatement(typescript_1.factory.createIdentifier('clearOnSuccess'), typescript_1.factory.createBlock([typescript_1.factory.createExpressionStatement(typescript_1.factory.createCallExpression(form_state_1.resetValuesName, undefined, []))], true), undefined),
        ];
    }
    return [];
};
exports.resetFunctionCheck = resetFunctionCheck;
const getValueIdentifier = (componentName, componentType) => {
    // For Boolean components like SwitchField, they need the full dot notation name
    // e.g. isChecked={options.enabled}
    if (typescript_type_map_1.FIELD_TYPE_TO_TYPESCRIPT_MAP[componentType] === typescript_1.SyntaxKind.BooleanKeyword) {
        return typescript_1.factory.createIdentifier(componentName);
    }
    return typescript_1.factory.createIdentifier(componentName.split('.')[0]);
};
