// @aws-amplify/graphql-directives/lib/directives/appsync/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsLambdaDirective = exports.AwsCognitoUserPoolsDirective = exports.AwsOidcDirective = exports.AwsIamDirective = exports.AwsApiKeyDirective = exports.AwsAuthDirective = exports.AwsSubscribeDirective = void 0;
var aws_subscribe_1 = require("./aws-subscribe");
Object.defineProperty(exports, "AwsSubscribeDirective", { enumerable: true, get: function () { return aws_subscribe_1.AwsSubscribeDirective; } });
var aws_auth_1 = require("./aws-auth");
Object.defineProperty(exports, "AwsAuthDirective", { enumerable: true, get: function () { return aws_auth_1.AwsAuthDirective; } });
var aws_api_key_1 = require("./aws-api-key");
Object.defineProperty(exports, "AwsApiKeyDirective", { enumerable: true, get: function () { return aws_api_key_1.AwsApiKeyDirective; } });
var aws_iam_1 = require("./aws-iam");
Object.defineProperty(exports, "AwsIamDirective", { enumerable: true, get: function () { return aws_iam_1.AwsIamDirective; } });
var aws_oidc_1 = require("./aws-oidc");
Object.defineProperty(exports, "AwsOidcDirective", { enumerable: true, get: function () { return aws_oidc_1.AwsOidcDirective; } });
var aws_cognito_user_pools_1 = require("./aws-cognito-user-pools");
Object.defineProperty(exports, "AwsCognitoUserPoolsDirective", { enumerable: true, get: function () { return aws_cognito_user_pools_1.AwsCognitoUserPoolsDirective; } });
var aws_lambda_1 = require("./aws-lambda");
Object.defineProperty(exports, "AwsLambdaDirective", { enumerable: true, get: function () { return aws_lambda_1.AwsLambdaDirective; } });
