// @aws-amplify/graphql-model-transformer/lib/resolvers/generators/dynamodb-vtl-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoDBModelVTLGenerator = void 0;
const dynamodb_1 = require("../dynamodb");
class DynamoDBModelVTLGenerator {
    generateUpdateRequestTemplate(config, ctx) {
        return (0, dynamodb_1.generateUpdateRequestTemplate)(config.modelName, config.isSyncEnabled);
    }
    generateCreateRequestTemplate(config, ctx) {
        return (0, dynamodb_1.generateCreateRequestTemplate)(config.modelName, config.modelIndexFields);
    }
    generateCreateInitSlotTemplate(config, initializeIdField) {
        return (0, dynamodb_1.generateCreateInitSlotTemplate)(config.modelConfig, initializeIdField);
    }
    generateDeleteRequestTemplate(config, ctx) {
        return (0, dynamodb_1.generateDeleteRequestTemplate)(config.modelName, config.isSyncEnabled);
    }
    generateUpdateInitSlotTemplate(config) {
        return (0, dynamodb_1.generateUpdateInitSlotTemplate)(config.modelConfig);
    }
    generateGetRequestTemplate(config, ctx) {
        return (0, dynamodb_1.generateGetRequestTemplate)();
    }
    generateGetResponseTemplate(config) {
        return (0, dynamodb_1.generateGetResponseTemplate)(config.isSyncEnabled);
    }
    generateListRequestTemplate(config, ctx) {
        return (0, dynamodb_1.generateListRequestTemplate)();
    }
    generateSyncRequestTemplate(config) {
        return (0, dynamodb_1.generateSyncRequestTemplate)();
    }
    generateSubscriptionRequestTemplate() {
        return (0, dynamodb_1.generateSubscriptionRequestTemplate)();
    }
    generateSubscriptionResponseTemplate() {
        return (0, dynamodb_1.generateSubscriptionResponseTemplate)();
    }
    generateDefaultResponseMappingTemplate(config) {
        return (0, dynamodb_1.generateDefaultResponseMappingTemplate)(config.isSyncEnabled, config.mutation);
    }
}
exports.DynamoDBModelVTLGenerator = DynamoDBModelVTLGenerator;
