// @aws-amplify/graphql-model-transformer/lib/resolvers/generators/rds-vtl-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RDSModelVTLGenerator = void 0;
const rds_1 = require("../rds");
const dynamodb_1 = require("../dynamodb");
class RDSModelVTLGenerator {
    generateUpdateRequestTemplate(config, ctx) {
        var _a;
        return (0, rds_1.generateLambdaUpdateRequestTemplate)(config.modelName, config.operationName, (_a = config.modelIndexFields) !== null && _a !== void 0 ? _a : ['id'], ctx);
    }
    generateCreateRequestTemplate(config, ctx) {
        return (0, rds_1.generateLambdaCreateRequestTemplate)(config.modelName, config.operationName, ctx);
    }
    generateCreateInitSlotTemplate(config, initializeIdField) {
        return (0, rds_1.generateCreateInitSlotTemplate)(config.modelConfig, initializeIdField);
    }
    generateDeleteRequestTemplate(config, ctx) {
        var _a;
        return (0, rds_1.generateLambdaDeleteRequestTemplate)(config.modelName, config.operationName, (_a = config.modelIndexFields) !== null && _a !== void 0 ? _a : ['id'], ctx);
    }
    generateUpdateInitSlotTemplate(config) {
        return (0, rds_1.generateUpdateInitSlotTemplate)(config.modelConfig);
    }
    generateGetRequestTemplate(config, ctx) {
        return (0, rds_1.generateLambdaRequestTemplate)(config.modelName, config.operation, config.operationName, ctx, true);
    }
    generateGetResponseTemplate(config) {
        return (0, rds_1.generateGetLambdaResponseTemplate)(false);
    }
    generateListRequestTemplate(config, ctx) {
        return (0, rds_1.generateLambdaListRequestTemplate)(config.modelName, config.operation, config.operationName, ctx);
    }
    generateSyncRequestTemplate(config) {
        return (0, rds_1.generateDefaultLambdaResponseMappingTemplate)(false);
    }
    generateSubscriptionRequestTemplate() {
        return (0, dynamodb_1.generateSubscriptionRequestTemplate)();
    }
    generateSubscriptionResponseTemplate() {
        return (0, dynamodb_1.generateSubscriptionResponseTemplate)();
    }
    generateDefaultResponseMappingTemplate(config) {
        return (0, rds_1.generateDefaultLambdaResponseMappingTemplate)(false, config.mutation);
    }
}
exports.RDSModelVTLGenerator = RDSModelVTLGenerator;
