// @aws-amplify/codegen-ui/dist/lib/utils/form-to-component/helpers/map-element-children.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapElementChildren = void 0;
function mapOptions(elementName, element) {
    const options = element.valueMappings.values.map(({ displayValue, value }, index) => {
        const optionType = element.componentType === 'RadioGroupField' ? 'Radio' : 'option';
        const option = {
            name: `${elementName}${optionType}${index}`,
            componentType: optionType,
            properties: {
                children: displayValue !== null && displayValue !== void 0 ? displayValue : value,
                value,
            },
        };
        if (element.componentType === 'SelectField' &&
            typeof value === 'object' &&
            'value' in value &&
            value.value === element.defaultValue) {
            option.properties.selected = { value: true };
        }
        return option;
    });
    return { children: options };
}
function mapElementChildren(elementName, element) {
    switch (element.componentType) {
        case 'SelectField':
        case 'RadioGroupField':
            return mapOptions(elementName, element);
        default:
            return { children: [] };
    }
}
exports.mapElementChildren = mapElementChildren;
