// @aws-amplify/graphql-relational-transformer/lib/graphql-has-many-transformer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HasManyTransformer = void 0;
const graphql_directives_1 = require("@aws-amplify/graphql-directives");
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const graphql_1 = require("graphql");
const graphql_transformer_common_1 = require("graphql-transformer-common");
const immer_1 = __importDefault(require("immer"));
const has_many_directive_transformer_factory_1 = require("./has-many/has-many-directive-transformer-factory");
const schema_1 = require("./schema");
const utils_1 = require("./utils");
class HasManyTransformer extends graphql_transformer_core_1.TransformerPluginBase {
    constructor() {
        super('amplify-has-many-transformer', graphql_directives_1.HasManyDirective.definition);
        this.directiveList = [];
        this.field = (parent, definition, directive, context) => {
            const directiveWrapped = new graphql_transformer_core_1.DirectiveWrapper(directive);
            const args = directiveWrapped.getArguments({
                directiveName: graphql_directives_1.HasManyDirective.name,
                object: parent,
                field: definition,
                directive,
                limit: graphql_directives_1.HasManyDirective.defaults.limit,
            }, (0, graphql_transformer_core_1.generateGetArgumentsInput)(context.transformParameters));
            validate(args, context);
            this.directiveList.push(args);
        };
        this.mutateSchema = (context) => {
            const resultDoc = (0, immer_1.default)(context.inputDocument, (draftDoc) => {
                var _a, _b;
                const connectingFieldsMap = new Map();
                const filteredDefs = (_a = draftDoc === null || draftDoc === void 0 ? void 0 : draftDoc.definitions) === null || _a === void 0 ? void 0 : _a.filter((def) => def.kind === 'ObjectTypeDefinition' || def.kind === 'ObjectTypeExtension');
                const objectDefs = filteredDefs;
                objectDefs === null || objectDefs === void 0 ? void 0 : objectDefs.forEach((def) => {
                    var _a;
                    const filteredFields = (_a = def === null || def === void 0 ? void 0 : def.fields) === null || _a === void 0 ? void 0 : _a.filter((field) => { var _a; return (_a = field === null || field === void 0 ? void 0 : field.directives) === null || _a === void 0 ? void 0 : _a.some((dir) => dir.name.value === graphql_directives_1.HasManyDirective.name); });
                    filteredFields === null || filteredFields === void 0 ? void 0 : filteredFields.forEach((field) => {
                        const baseFieldType = (0, graphql_transformer_common_1.getBaseType)(field.type);
                        const connectionAttributeName = (0, utils_1.getConnectionAttributeName)(context.transformParameters, def.name.value, field.name.value, (0, utils_1.getObjectPrimaryKey)(def).name.value);
                        const newField = (0, graphql_transformer_common_1.makeField)(connectionAttributeName, [], (0, graphql_transformer_common_1.isNonNullType)(field.type) ? (0, graphql_transformer_common_1.makeNonNullType)((0, graphql_transformer_common_1.makeNamedType)('ID')) : (0, graphql_transformer_common_1.makeNamedType)('ID'), []);
                        const sortKeyFields = (0, schema_1.convertSortKeyFieldsToSortKeyConnectionFields)((0, schema_1.getSortKeyFieldsNoContext)(def), def, field);
                        const allNewFields = [newField, ...sortKeyFields];
                        connectingFieldsMap.set(baseFieldType, allNewFields);
                    });
                });
                (_b = objectDefs === null || objectDefs === void 0 ? void 0 : objectDefs.filter((def) => connectingFieldsMap.has(def.name.value))) === null || _b === void 0 ? void 0 : _b.forEach((def) => {
                    const fieldsToAdd = connectingFieldsMap.get(def.name.value);
                    if (fieldsToAdd) {
                        (0, schema_1.addFieldsToDefinition)(def, fieldsToAdd);
                    }
                });
            });
            return resultDoc;
        };
        this.prepare = (context) => {
            this.directiveList.forEach((config) => {
                const modelName = config.object.name.value;
                const dbType = (0, graphql_transformer_core_1.getStrategyDbTypeFromModel)(context, modelName);
                const dataSourceBasedTransformer = (0, has_many_directive_transformer_factory_1.getHasManyDirectiveTransformer)(dbType, config);
                dataSourceBasedTransformer.prepare(context, config);
            });
        };
        this.transformSchema = (ctx) => {
            const context = ctx;
            for (const config of this.directiveList) {
                const dbType = (0, graphql_transformer_core_1.getStrategyDbTypeFromTypeNode)(config.field.type, context);
                const dataSourceBasedTransformer = (0, has_many_directive_transformer_factory_1.getHasManyDirectiveTransformer)(dbType, config);
                dataSourceBasedTransformer.transformSchema(ctx, config);
                (0, schema_1.ensureHasManyConnectionField)(config, context);
                (0, schema_1.extendTypeWithConnection)(config, context);
            }
        };
        this.generateResolvers = (ctx) => {
            const context = ctx;
            for (const config of this.directiveList) {
                if (!ctx.transformParameters.allowGen1Patterns) {
                    const { field, object } = config;
                    const modelName = object.name.value;
                    const fieldName = field.name.value;
                    if (field.type.kind === graphql_1.Kind.NON_NULL_TYPE) {
                        aws_cdk_lib_1.Annotations.of(ctx.api).addWarning(`@${graphql_directives_1.HasManyDirective.name} on required fields is deprecated. Modify ${modelName}.${fieldName} to be optional. This functionality will be removed in the next major release.`);
                    }
                    if (config.fields) {
                        aws_cdk_lib_1.Annotations.of(ctx.api).addWarning(`fields argument on @${graphql_directives_1.HasManyDirective.name} is deprecated. Modify ${modelName}.${fieldName} to use references instead. This functionality will be removed in the next major release.`);
                    }
                }
                const dbType = (0, graphql_transformer_core_1.getStrategyDbTypeFromTypeNode)(config.field.type, context);
                const dataSourceBasedTransformer = (0, has_many_directive_transformer_factory_1.getHasManyDirectiveTransformer)(dbType, config);
                dataSourceBasedTransformer.generateResolvers(ctx, config);
            }
        };
    }
}
exports.HasManyTransformer = HasManyTransformer;
const validate = (config, ctx) => {
    var _a, _b;
    const { field } = config;
    if (!(0, graphql_transformer_common_1.isListType)(field.type)) {
        throw new graphql_transformer_core_1.InvalidDirectiveError(`@${graphql_directives_1.HasManyDirective.name} must be used with a list. Use @hasOne for non-list types.`);
    }
    let dbType;
    try {
        dbType = (0, graphql_transformer_core_1.getStrategyDbTypeFromTypeNode)(field.type, ctx);
    }
    catch (_c) {
        throw new graphql_transformer_core_1.InvalidDirectiveError(`Object type ${(_b = (_a = field.type) === null || _a === void 0 ? void 0 : _a.name.value) !== null && _b !== void 0 ? _b : field.name} must be annotated with @model.`);
    }
    config.relatedType = (0, utils_1.getRelatedType)(config, ctx);
    const dataSourceBasedTransformer = (0, has_many_directive_transformer_factory_1.getHasManyDirectiveTransformer)(dbType, config);
    dataSourceBasedTransformer.validate(ctx, config);
    (0, utils_1.validateModelDirective)(config);
    config.connectionFields = [];
    (0, utils_1.validateRelatedModelDirective)(config);
    (0, utils_1.validateDisallowedDataStoreRelationships)(config, ctx);
};
