// @aws-amplify/graphql-model-transformer/lib/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWS_IAM_DIRECTIVE = exports.API_KEY_DIRECTIVE = exports.OPERATION_KEY = exports.ATTRIBUTE_TYPES = exports.BOOLEAN_FUNCTIONS = exports.FLOAT_FUNCTIONS = exports.INT_FUNCTIONS = exports.ID_FUNCTIONS = exports.STRING_FUNCTIONS = exports.SUBSCRIPTION_BOOLEAN_CONDITIONS = exports.SUBSCRIPTION_FLOAT_CONDITIONS = exports.SUBSCRIPTION_INT_CONDITIONS = exports.SUBSCRIPTION_ID_CONDITIONS = exports.SUBSCRIPTION_STRING_CONDITIONS = exports.SIZE_CONDITIONS = exports.BOOLEAN_CONDITIONS = exports.FLOAT_CONDITIONS = exports.INT_CONDITIONS = exports.ID_CONDITIONS = exports.STRING_CONDITIONS = void 0;
exports.STRING_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'contains', 'notContains', 'between', 'beginsWith'];
exports.ID_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'contains', 'notContains', 'between', 'beginsWith'];
exports.INT_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'between'];
exports.FLOAT_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'between'];
exports.BOOLEAN_CONDITIONS = ['ne', 'eq'];
exports.SIZE_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'between'];
exports.SUBSCRIPTION_STRING_CONDITIONS = [
    'ne',
    'eq',
    'le',
    'lt',
    'ge',
    'gt',
    'contains',
    'notContains',
    'between',
    'beginsWith',
    'in',
    'notIn',
];
exports.SUBSCRIPTION_ID_CONDITIONS = [
    'ne',
    'eq',
    'le',
    'lt',
    'ge',
    'gt',
    'contains',
    'notContains',
    'between',
    'beginsWith',
    'in',
    'notIn',
];
exports.SUBSCRIPTION_INT_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'between', 'in', 'notIn'];
exports.SUBSCRIPTION_FLOAT_CONDITIONS = ['ne', 'eq', 'le', 'lt', 'ge', 'gt', 'between', 'in', 'notIn'];
exports.SUBSCRIPTION_BOOLEAN_CONDITIONS = ['ne', 'eq'];
exports.STRING_FUNCTIONS = new Set(['attributeExists', 'attributeType', 'size']);
exports.ID_FUNCTIONS = new Set(['attributeExists', 'attributeType', 'size']);
exports.INT_FUNCTIONS = new Set(['attributeExists', 'attributeType']);
exports.FLOAT_FUNCTIONS = new Set(['attributeExists', 'attributeType']);
exports.BOOLEAN_FUNCTIONS = new Set(['attributeExists', 'attributeType']);
exports.ATTRIBUTE_TYPES = ['binary', 'binarySet', 'bool', 'list', 'map', 'number', 'numberSet', 'string', 'stringSet', '_null'];
exports.OPERATION_KEY = '__operation';
exports.API_KEY_DIRECTIVE = 'aws_api_key';
exports.AWS_IAM_DIRECTIVE = 'aws_iam';
