// @aws-amplify/platform-core/lib/parameter_path_conversions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterPathConversions = void 0;
const backend_identifier_conversions_js_1 = require("./backend_identifier_conversions.js");
const errors_1 = require("./errors");
const SHARED_SECRET = 'shared';
const RESOURCE_REFERENCE = 'resource_reference';
/**
 * Provides static methods for converting BackendIdentifier to parameter path strings
 */
class ParameterPathConversions {
    /**
     * Convert a BackendIdentifier to a parameter prefix.
     */
    // It's fine to ignore the rule here because the anti-static rule is to ban the static function which should use constructor
    // eslint-disable-next-line no-restricted-syntax
    static toParameterPrefix(backendId) {
        if (typeof backendId === 'object') {
            return getBackendParameterPrefix(backendId);
        }
        return getSharedParameterPrefix(backendId);
    }
    /**
     * Convert a BackendIdentifier to a parameter full path.
     */
    // It's fine to ignore the rule here because the anti-static rule is to ban the static function which should use constructor
    // eslint-disable-next-line no-restricted-syntax
    static toParameterFullPath(backendId, parameterName) {
        if (typeof backendId === 'object') {
            return getBackendParameterFullPath(backendId, parameterName);
        }
        return getSharedParameterFullPath(backendId, parameterName);
    }
    /**
     * Generate an SSM path for references to other backend resources
     */
    // It's fine to ignore the rule here because the anti-static rule is to ban the static function which should use constructor
    // eslint-disable-next-line no-restricted-syntax
    static toResourceReferenceFullPath(backendId, referenceName) {
        return `/amplify/${RESOURCE_REFERENCE}/${getBackendIdentifierPathPart(backendId)}/${referenceName}`;
    }
}
exports.ParameterPathConversions = ParameterPathConversions;
const getBackendParameterPrefix = (parts) => {
    return `/amplify/${getBackendIdentifierPathPart(parts)}`;
};
/**
 * Get a branch-specific parameter prefix.
 */
const getBackendIdentifierPathPart = (parts) => {
    // round trip the backend id through the stack name conversion to ensure we are applying the same sanitization to SSM paths
    const sanitizedBackendId = backend_identifier_conversions_js_1.BackendIdentifierConversions.fromStackName(backend_identifier_conversions_js_1.BackendIdentifierConversions.toStackName(parts));
    if (!sanitizedBackendId || !sanitizedBackendId.hash) {
        // this *should* never happen
        throw new errors_1.AmplifyFault('BackendIdConversionFault', {
            message: `Could not sanitize the backendId to construct the parameter path`,
        });
    }
    return `${sanitizedBackendId.namespace}/${sanitizedBackendId.name}-${sanitizedBackendId.type}-${sanitizedBackendId.hash}`;
};
/**
 * Get a branch-specific parameter full path.
 */
const getBackendParameterFullPath = (backendIdentifier, parameterName) => {
    return `${getBackendParameterPrefix(backendIdentifier)}/${parameterName}`;
};
/**
 * Get a shared parameter prefix.
 */
const getSharedParameterPrefix = (appId) => {
    return `/amplify/${SHARED_SECRET}/${appId}`;
};
/**
 * Get a shared parameter full path.
 */
const getSharedParameterFullPath = (appId, parameterName) => {
    return `${getSharedParameterPrefix(appId)}/${parameterName}`;
};
