// @aws-amplify/graphql-relational-transformer/lib/belongs-to/belongs-to-directive-ddb-fields-transformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BelongsToDirectiveDDBFieldsTransformer = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const generator_factory_1 = require("../resolver/generator-factory");
const utils_1 = require("../utils");
class BelongsToDirectiveDDBFieldsTransformer {
    constructor() {
        this.dbType = graphql_transformer_core_1.DDB_DB_TYPE;
        this.prepare = (context, config) => {
            if (config.relationType !== 'hasOne') {
                return;
            }
            const modelName = config.object.name.value;
            (0, utils_1.registerHasOneForeignKeyMappings)({
                transformParameters: context.transformParameters,
                resourceHelper: context.resourceHelper,
                thisTypeName: modelName,
                thisFieldName: config.field.name.value,
                relatedType: config.relatedType,
            });
        };
        this.transformSchema = (context, config) => {
            config.relatedTypeIndex = (0, utils_1.getRelatedTypeIndex)(config, context);
        };
        this.generateResolvers = (context, config) => {
            const generator = (0, generator_factory_1.getGenerator)(this.dbType);
            generator.makeBelongsToGetItemConnectionWithKeyResolver(config, context);
        };
        this.validate = (context, config) => {
            (0, utils_1.ensureFieldsArray)(config);
            config.fieldNodes = (0, utils_1.getFieldsNodes)(config, context);
        };
    }
}
exports.BelongsToDirectiveDDBFieldsTransformer = BelongsToDirectiveDDBFieldsTransformer;
