// @aws-amplify/codegen-ui/dist/lib/generate-form-definition/helpers/model-fields-configs.js
"use strict";
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapModelFieldsConfigs = exports.getFieldConfigFromModelField = exports.getFieldTypeMapKey = void 0;
const change_case_1 = require("change-case");
const check_support_1 = require("../../check-support");
const errors_1 = require("../../errors");
const field_type_map_1 = require("./field-type-map");
const isModelDataType = (field) => typeof field.dataType === 'object' && 'model' in field.dataType;
const isHasManyRelationship = (relationship) => { var _a; return (_a = ('isHasManyIndex' in relationship && relationship.isHasManyIndex)) !== null && _a !== void 0 ? _a : false; };
function extractCorrespondingKey({ thisModel, relatedModel, relationshipFieldName, }) {
    var _a;
    const { relationship: thisModelRelationship } = (_a = thisModel.fields[relationshipFieldName]) !== null && _a !== void 0 ? _a : {};
    if (thisModelRelationship) {
        if (isHasManyRelationship(thisModelRelationship)) {
            const correspondingFieldTuple = Object.values(relatedModel.fields).find(({ relationship }) => (relationship === null || relationship === void 0 ? void 0 : relationship.type) === 'HAS_MANY' && (relationship === null || relationship === void 0 ? void 0 : relationship.relatedModelFields.includes(relationshipFieldName)));
            if (correspondingFieldTuple) {
                const correspondingField = correspondingFieldTuple.relationship;
                if ((correspondingField === null || correspondingField === void 0 ? void 0 : correspondingField.type) === 'HAS_MANY') {
                    const indexOfKey = correspondingField.relatedModelFields.indexOf(relationshipFieldName);
                    if (indexOfKey !== -1) {
                        const relatedPrimaryKey = relatedModel.primaryKeys[indexOfKey];
                        if (relatedPrimaryKey) {
                            // secondary index on child of 1:m
                            return relatedPrimaryKey;
                        }
                    }
                }
            }
        }
        else {
            const modelRelationshipFieldTuple = Object.values(thisModel.fields)
                .filter(isModelDataType)
                .find(({ relationship: matchingFieldRelationship }) => {
                var _a;
                return ((matchingFieldRelationship === null || matchingFieldRelationship === void 0 ? void 0 : matchingFieldRelationship.type) === thisModelRelationship.type &&
                    (matchingFieldRelationship === null || matchingFieldRelationship === void 0 ? void 0 : matchingFieldRelationship.relatedModelName) === thisModelRelationship.relatedModelName &&
                    ((_a = matchingFieldRelationship === null || matchingFieldRelationship === void 0 ? void 0 : matchingFieldRelationship.associatedFields) === null || _a === void 0 ? void 0 : _a.includes(relationshipFieldName)));
            });
            if (modelRelationshipFieldTuple) {
                const modelRelationshipField = modelRelationshipFieldTuple.relationship;
                if (modelRelationshipField &&
                    'associatedFields' in modelRelationshipField &&
                    modelRelationshipField.associatedFields) {
                    const indexOfKey = modelRelationshipField.associatedFields.indexOf(relationshipFieldName);
                    if (indexOfKey !== -1) {
                        const relatedPrimaryKey = relatedModel.primaryKeys[indexOfKey];
                        if (relatedPrimaryKey) {
                            // index on parent of 1:1
                            return relatedPrimaryKey;
                        }
                    }
                }
            }
        }
    }
    throw new errors_1.InternalError(`Cannot find corresponding key for scalar relationship field ${relationshipFieldName}`);
}
function getFieldTypeMapKey(field) {
    var _a;
    if (typeof field.dataType === 'object' && 'enum' in field.dataType) {
        return 'Enum';
    }
    if ((typeof field.dataType === 'object' && 'model' in field.dataType) || ((_a = field.relationship) === null || _a === void 0 ? void 0 : _a.relatedModelName)) {
        return 'Relationship';
    }
    if (typeof field.dataType === 'object' && 'nonModel' in field.dataType) {
        return 'NonModel';
    }
    return field.dataType;
}
exports.getFieldTypeMapKey = getFieldTypeMapKey;
function getModelDisplayValue({ model, modelName, keyToMap, }) {
    var _a;
    const concatArray = [];
    const { primaryKeys } = model;
    const scalarNonReadableFields = new Set(['ID', 'AWSJSON', 'AWSTime', 'AWSDate', 'AWSDateTime']);
    const firstReadableNonKeyField = (_a = Object.entries(model.fields).find(([fieldName, fieldObject]) => !primaryKeys.includes(fieldName) &&
        !fieldObject.relationship &&
        !(typeof fieldObject.dataType === 'string' && scalarNonReadableFields.has(fieldObject.dataType)) &&
        !(typeof fieldObject.dataType === 'object' && 'nonModel' in fieldObject.dataType) &&
        !isModelDataType(fieldObject))) === null || _a === void 0 ? void 0 : _a[0];
    if (firstReadableNonKeyField) {
        concatArray.push({
            bindingProperties: { property: modelName, field: firstReadableNonKeyField },
        }, {
            value: ' - ',
        });
    }
    if (keyToMap) {
        concatArray.push({
            bindingProperties: { property: modelName, field: keyToMap },
        });
    }
    else {
        primaryKeys.forEach((key, index) => {
            concatArray.push({
                bindingProperties: { property: modelName, field: key },
            });
            if (index !== primaryKeys.length - 1) {
                concatArray.push({
                    value: '-',
                });
            }
        });
    }
    return concatArray.length === 1 ? { ...concatArray[0], isDefault: true } : { concat: concatArray, isDefault: true };
}
function getValueMappings({ dataTypeName, fieldName, field, enums, allModels, }) {
    // if enum
    if (typeof field.dataType === 'object' && 'enum' in field.dataType) {
        const fieldEnums = enums[field.dataType.enum];
        if (!fieldEnums) {
            throw new errors_1.InvalidInputError(`Values could not be found for enum ${field.dataType.enum}`);
        }
        return {
            values: fieldEnums.values.map((value) => ({
                displayValue: { value: (0, change_case_1.sentenceCase)(value) ? (0, change_case_1.sentenceCase)(value) : value, isDefault: true },
                value: { value },
            })),
        };
    }
    // if relationship
    if (field.relationship) {
        const modelName = field.relationship.relatedModelName;
        const relatedModel = allModels[modelName];
        const isModelType = isModelDataType(field);
        // if model, store all keys; else, store corresponding primary key
        const keys = isModelType
            ? relatedModel.primaryKeys
            : [
                extractCorrespondingKey({
                    thisModel: allModels[dataTypeName],
                    relatedModel,
                    relationshipFieldName: fieldName,
                }),
            ];
        const values = keys.map((key) => ({
            value: { bindingProperties: { property: modelName, field: key } },
        }));
        values[0].displayValue = getModelDisplayValue({
            model: relatedModel,
            modelName,
            keyToMap: isModelType ? undefined : keys[0],
        });
        return {
            values,
            bindingProperties: { [modelName]: { type: 'Data', bindingProperties: { model: modelName } } },
        };
    }
    return undefined;
}
function getFieldConfigFromModelField({ dataTypeName, fieldName, field, dataSchema, setReadOnly, featureFlags, }) {
    var _a;
    const fieldTypeMapKey = getFieldTypeMapKey(field);
    let { defaultComponent } = field_type_map_1.FIELD_TYPE_MAP[fieldTypeMapKey];
    // we are rolling out the switch to NumberField with relationship support
    if (!(featureFlags === null || featureFlags === void 0 ? void 0 : featureFlags.isRelationshipSupported) && fieldTypeMapKey === 'AWSTimestamp') {
        defaultComponent = 'DateTimeField';
    }
    // When the relationship is many to many, set data type to the actual related model instead of the join table
    // if (field.relationship && field.relationship.type === 'HAS_MANY' && field.relationship.relatedJoinTableName) {
    //   const dataType = field.dataType as { model: string };
    //   dataType.model = field.relationship.relatedModelName;
    // }
    let { dataType } = field;
    if (field.relationship && isModelDataType(field) && ((_a = dataSchema.models[field.dataType.model]) === null || _a === void 0 ? void 0 : _a.isJoinTable)) {
        dataType = { model: field.relationship.relatedModelName };
    }
    let { readOnly } = field;
    if (setReadOnly) {
        readOnly = true;
    }
    const config = {
        label: (0, change_case_1.sentenceCase)(fieldName),
        dataType,
        inputType: {
            type: defaultComponent,
            required: field.required,
            readOnly,
            name: fieldName,
            value: fieldName,
            isArray: field.isArray,
        },
    };
    if (field.relationship) {
        config.relationship = field.relationship;
        config.inputType.placeholder = `Search ${field.relationship.relatedModelName}`;
    }
    const valueMappings = getValueMappings({
        dataTypeName,
        fieldName,
        field,
        enums: dataSchema.enums,
        allModels: dataSchema.models,
    });
    if (valueMappings) {
        config.inputType.valueMappings = valueMappings;
    }
    return config;
}
exports.getFieldConfigFromModelField = getFieldConfigFromModelField;
/**
 * Impure function that adds fields from DataStore to temporary util object, modelFieldsConfigs
 * and to the formDefinition
 */
function mapModelFieldsConfigs({ dataTypeName, formDefinition, dataSchema, formActionType, featureFlags, }) {
    const modelFieldsConfigs = {};
    const model = dataSchema.models[dataTypeName];
    if (!model) {
        throw new errors_1.InvalidInputError(`Model ${dataTypeName} not found`);
    }
    Object.entries(model.fields).forEach(([fieldName, field]) => {
        const isAutoExcludedField = field.readOnly ||
            (fieldName === 'id' && field.dataType === 'ID' && field.required) ||
            !(0, check_support_1.checkIsSupportedAsFormField)(field, featureFlags) ||
            (field.relationship &&
                !isModelDataType(field) &&
                // if index overlaps with that of BELONGS_TO, the model field will establish bidirectionality automatically
                !('isHasManyIndex' in field.relationship &&
                    field.relationship.isHasManyIndex &&
                    field.relationship.type === 'HAS_ONE'));
        if (!isAutoExcludedField) {
            formDefinition.elementMatrix.push([fieldName]);
        }
        const isPrimaryKey = model.primaryKeys.includes(fieldName);
        modelFieldsConfigs[fieldName] = getFieldConfigFromModelField({
            dataTypeName,
            fieldName,
            field,
            dataSchema,
            setReadOnly: isPrimaryKey && formActionType === 'update',
            featureFlags,
        });
    });
    return modelFieldsConfigs;
}
exports.mapModelFieldsConfigs = mapModelFieldsConfigs;
