// @aws-amplify/backend-cli/lib/form-generation/form_generation_handler.js
import { createLocalGraphqlFormGenerator } from '@aws-amplify/form-generator';
import { createGraphqlDocumentGenerator } from '@aws-amplify/model-generator';
import { printer } from '@aws-amplify/cli-core';
/**
 * Creates a handler for FormGeneration
 */
export class FormGenerationHandler {
    formGenParams;
    /**
     * Instantiates the handler
     */
    constructor(formGenParams) {
        this.formGenParams = formGenParams;
    }
    generate = async (params) => {
        const { backendIdentifier, modelsOutDir, uiOutDir, apiUrl, modelsFilter } = params;
        const { awsClientProvider } = this.formGenParams;
        const graphqlClientGenerator = createGraphqlDocumentGenerator({
            backendIdentifier,
            awsClientProvider,
        });
        const modelsResult = await graphqlClientGenerator.generateModels({
            targetFormat: 'typescript',
        });
        const { filesWritten: modelsFilesWritten } = await modelsResult.writeToDirectory(modelsOutDir);
        this.logMessages(modelsFilesWritten);
        const localFormGenerator = createLocalGraphqlFormGenerator({
            introspectionSchemaUrl: apiUrl,
            graphqlModelDirectoryPath: './graphql',
        });
        const result = await localFormGenerator.generateForms({
            models: modelsFilter,
        });
        const { filesWritten: uiFilesWritten } = await result.writeToDirectory(uiOutDir);
        this.logMessages(uiFilesWritten);
    };
    logMessages = (filesWritten) => {
        filesWritten.forEach((file) => {
            printer.log(`File written: ${file}`);
        });
    };
}
