// @aws-amplify/backend/lib/engine/backend-secret/lambda/backend_secret_fetcher.js
import { getSecretClient, } from '@aws-amplify/backend-secret';
import { randomUUID } from 'node:crypto';
const secretClient = getSecretClient();
/**
 * Entry point for the lambda-backend custom resource to retrieve a backend secret.
 */
export const handler = async (event) => {
    console.info(`Received '${event.RequestType}' event`);
    const physicalId = event.RequestType === 'Create' ? randomUUID() : event.PhysicalResourceId;
    let data = undefined;
    if (event.RequestType === 'Update' || event.RequestType === 'Create') {
        const val = await handleCreateUpdateEvent(secretClient, event);
        data = {
            secretValue: val,
        };
    }
    return {
        RequestId: event.RequestId,
        LogicalResourceId: event.LogicalResourceId,
        PhysicalResourceId: physicalId,
        Data: data,
        StackId: event.StackId,
        NoEcho: true,
        Status: 'SUCCESS',
    };
};
/**
 * Handles create/update event for the secret custom resource.
 */
export const handleCreateUpdateEvent = async (secretClient, event) => {
    const props = event.ResourceProperties;
    let secret;
    try {
        const resp = await secretClient.getSecret({
            namespace: props.namespace,
            name: props.name,
            type: props.type,
        }, {
            name: props.secretName,
        });
        secret = resp?.value;
    }
    catch (err) {
        const secretErr = err;
        if (secretErr.httpStatusCode && secretErr.httpStatusCode >= 500) {
            throw new Error(`Failed to retrieve backend secret '${props.secretName}' for '${props.namespace}/${props.name}'. Reason: ${JSON.stringify(err)}`);
        }
    }
    // if the secret is not available in branch path, try retrieving it at the app-level.
    if (!secret) {
        try {
            const resp = await secretClient.getSecret(props.namespace, {
                name: props.secretName,
            });
            secret = resp?.value;
        }
        catch (err) {
            throw new Error(`Failed to retrieve backend secret '${props.secretName}' for '${props.namespace}'. Reason: ${JSON.stringify(err)}`);
        }
    }
    if (!secret) {
        throw new Error(`Unable to find backend secret for backend '${props.namespace}', branch '${props.name}', name '${props.secretName}'`);
    }
    return secret;
};
