// @aws-amplify/codegen-ui-react/dist/lib/react-theme-studio-template-renderer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactThemeStudioTemplateRenderer = void 0;
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
const os_1 = require("os");
const typescript_1 = require("typescript");
const codegen_ui_1 = require("@aws-amplify/codegen-ui");
const react_render_config_1 = require("./react-render-config");
const imports_1 = require("./imports");
const react_output_manager_1 = require("./react-output-manager");
const react_studio_template_renderer_helper_1 = require("./react-studio-template-renderer-helper");
class ReactThemeStudioTemplateRenderer extends codegen_ui_1.StudioTemplateRenderer {
    constructor(theme, renderConfig, options) {
        super(theme, new react_output_manager_1.ReactOutputManager(), renderConfig);
        this.importCollection = new imports_1.ImportCollection();
        this.renderConfig = {
            ...react_studio_template_renderer_helper_1.defaultRenderConfig,
            ...renderConfig,
        };
        this.fileName = `${this.component.name}.${(0, react_render_config_1.scriptKindToFileExtensionNonReact)(this.renderConfig.script)}`;
        this.options = options;
    }
    renderComponentInternal() {
        const { printer, file } = (0, react_studio_template_renderer_helper_1.buildPrinter)(this.fileName, this.renderConfig);
        const renderedImports = this.buildImports().map((importStatement) => printer.printNode(typescript_1.EmitHint.Unspecified, importStatement, file));
        const renderedFunction = printer.printNode(typescript_1.EmitHint.Unspecified, this.buildTheme(), file);
        const componentText = ['/* eslint-disable */', ...renderedImports, renderedFunction].join(os_1.EOL);
        const { componentText: transpiledComponentText, declaration } = (0, react_studio_template_renderer_helper_1.transpile)(componentText, this.renderConfig);
        return {
            componentText: transpiledComponentText,
            declaration,
            renderComponentToFilesystem: async (outputPath) => {
                await this.renderComponentToFilesystem(transpiledComponentText)(this.fileName)(outputPath);
                if (declaration) {
                    await this.renderComponentToFilesystem(declaration)((0, react_studio_template_renderer_helper_1.getDeclarationFilename)(this.fileName))(outputPath);
                }
            },
        };
    }
    /**
     * Exposing an additional method to allow rendering the theme json object
     * from the API response type.
     */
    renderThemeJson() {
        const { printer, file } = (0, react_studio_template_renderer_helper_1.buildPrinter)(this.fileName, this.renderConfig);
        const themeJson = printer.printNode(typescript_1.EmitHint.Unspecified, this.buildThemeObject(), file);
        // prettier fails if we don't provide a valid statement, so wrapping in a statement and removing after formatting
        return (0, react_studio_template_renderer_helper_1.formatCode)(`const a = ${themeJson};`).replace('const a =', '').replace(';', '').trim();
    }
    /*
     * import { createTheme } from "@aws-amplify/ui-react";
     */
    buildImports() {
        var _a;
        this.importCollection.addMappedImport(imports_1.ImportValue.CREATE_THEME);
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.renderDefaultTheme) {
            this.importCollection.addMappedImport(imports_1.ImportValue.DEFAULT_THEME);
        }
        return this.importCollection.buildImportStatements(true);
    }
    /*
     * export default createTheme({ ... });
     */
    buildTheme() {
        var _a;
        return typescript_1.factory.createExportAssignment(undefined, undefined, undefined, typescript_1.factory.createCallExpression(typescript_1.factory.createIdentifier('createTheme'), undefined, [
            ((_a = this.options) === null || _a === void 0 ? void 0 : _a.renderDefaultTheme)
                ? typescript_1.factory.createIdentifier(imports_1.ImportValue.DEFAULT_THEME)
                : this.buildThemeObject(),
        ]));
    }
    /*
     * {
     *   id: '123',
     *   name: 'MyTheme',
     *   tokens: {},
     *   overrides: {},
     * }
     */
    buildThemeObject() {
        return typescript_1.factory.createObjectLiteralExpression([
            typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier('name'), typescript_1.factory.createStringLiteral(this.component.name)),
        ]
            .concat(this.splitAndBuildThemeValues(this.component.values))
            .concat(this.buildThemeOverrides(this.component.overrides)), true);
    }
    splitAndBuildThemeValues(values) {
        return values.map(({ key, value }) => {
            if (key !== 'breakpoints') {
                return typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier(key), this.buildThemeValue(value));
            }
            return typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier(key), this.buildThemeBreakpointValue(value));
        });
    }
    /* Removes children and value (needed for smithy) from theme values json
     *
     * tokens: {
     *   components: {
     *     alert: {
     *       backgroundcolor: \\"hsl(210, 5%, 90%)\\",
     * ...
     */
    buildThemeValues(values) {
        return values.map(({ key, value }) => typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier(key), this.buildThemeValue(value)));
    }
    buildThemeValue(themeValue) {
        const { children, value } = themeValue;
        if (children) {
            return typescript_1.factory.createObjectLiteralExpression(this.buildThemeValues(children));
        }
        if (value) {
            return typescript_1.factory.createStringLiteral(value);
        }
        throw new codegen_ui_1.InvalidInputError(`Invalid theme value: ${JSON.stringify(value)}`);
    }
    /* Removes children and value (needed for smithy) from theme values json
     *
     * breakpoints: {
     *   values: {
     *     base: 0,
     *     small: 480,
     *     ...
     *   }
     *   defaultBreakpoint: "base",
     * ...
     */
    buildThemeBreakpointValues(values) {
        return values.map(({ key, value }) => typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier(key), this.buildThemeBreakpointValue(value)));
    }
    buildThemeBreakpointValue(themeValue) {
        const { children, value } = themeValue;
        if (children) {
            if (children[0].key === 'value') {
                return this.buildThemeBreakpointValue({ value: children[0].value.value });
            }
            return typescript_1.factory.createObjectLiteralExpression(this.buildThemeBreakpointValues(children));
        }
        if (value) {
            if (/^[0-9]+$/.test(value)) {
                return typescript_1.factory.createNumericLiteral(parseInt(value, 10));
            }
            return typescript_1.factory.createStringLiteral(value);
        }
        throw new codegen_ui_1.InvalidInputError(`Invalid theme value: ${JSON.stringify(value)}`);
    }
    /* builds special case theme value overrides becuase it is an array
     *
     * overrides: [
     *   {
     *     colorMode: \\"dark\\",
     *     tokens: {
     *       colors: { black: { value: \\"#fff\\" }, white: { value: \\"#000\\" } },
     *     },
     *   },
     * ],
     */
    buildThemeOverrides(overrides) {
        if (overrides === undefined) {
            return [];
        }
        return typescript_1.factory.createPropertyAssignment(typescript_1.factory.createIdentifier('overrides'), typescript_1.factory.createArrayLiteralExpression([typescript_1.factory.createObjectLiteralExpression(this.buildThemeValues(overrides), true)], false));
    }
    validateSchema(theme) {
        (0, codegen_ui_1.validateThemeSchema)(theme);
    }
}
__decorate([
    codegen_ui_1.handleCodegenErrors
], ReactThemeStudioTemplateRenderer.prototype, "renderThemeJson", null);
exports.ReactThemeStudioTemplateRenderer = ReactThemeStudioTemplateRenderer;
