// graphql-transformer-common/lib/ResolverResourceIDs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolverResourceIDs = void 0;
const util_1 = require("./util");
class ResolverResourceIDs {
    static DynamoDBCreateResolverResourceID(typeName) {
        return `Create${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static DynamoDBUpdateResolverResourceID(typeName) {
        return `Update${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static DynamoDBDeleteResolverResourceID(typeName) {
        return `Delete${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static DynamoDBGetResolverResourceID(typeName) {
        return `Get${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static DynamoDBListResolverResourceID(typeName) {
        return `List${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static ElasticsearchSearchResolverResourceID(typeName) {
        return `Search${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static SyncResolverResourceID(typeName) {
        return `Sync${(0, util_1.resourceName)(typeName)}Resolver`;
    }
    static ResolverResourceID(typeName, fieldName) {
        return `${(0, util_1.resourceName)(`${typeName}${fieldName}`)}Resolver`;
    }
}
exports.ResolverResourceIDs = ResolverResourceIDs;
